// node-opcua-pki/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/node-opcua-pki/lib/index.ts
var lib_exports = {};
__export(lib_exports, {
  CertificateAuthority: () => CertificateAuthority,
  CertificateManager: () => CertificateManager,
  CertificateManagerState: () => CertificateManagerState,
  Subject: () => import_node_opcua_crypto.Subject,
  VerificationStatus: () => VerificationStatus,
  adjustApplicationUri: () => adjustApplicationUri,
  adjustDate: () => adjustDate,
  convertPFXtoPEM: () => convertPFXtoPEM,
  createPFX: () => createPFX,
  dumpPFX: () => dumpPFX,
  extractAllFromPFX: () => extractAllFromPFX,
  extractCACertificatesFromPFX: () => extractCACertificatesFromPFX,
  extractCertificateFromPFX: () => extractCertificateFromPFX,
  extractPrivateKeyFromPFX: () => extractPrivateKeyFromPFX,
  findIssuerCertificateInChain: () => findIssuerCertificateInChain,
  install_prerequisite: () => install_prerequisite,
  quote: () => quote
});
module.exports = __toCommonJS(lib_exports);

// packages/node-opcua-pki/lib/ca/certificate_authority.ts
var import_node_assert6 = __toESM(require("assert"));
var import_node_fs6 = __toESM(require("fs"));
var import_node_path5 = __toESM(require("path"));
var import_chalk5 = __toESM(require("chalk"));
var import_node_opcua_crypto2 = require("node-opcua-crypto");

// packages/node-opcua-pki/lib/toolbox/common.ts
var import_node_assert = __toESM(require("assert"));
function quote(str) {
  return `"${str || ""}"`;
}
function adjustDate(params) {
  (0, import_node_assert.default)(params instanceof Object);
  params.startDate = params.startDate || /* @__PURE__ */ new Date();
  (0, import_node_assert.default)(params.startDate instanceof Date);
  params.validity = params.validity || 365;
  params.endDate = new Date(params.startDate.getTime());
  params.endDate.setDate(params.startDate.getDate() + params.validity);
  (0, import_node_assert.default)(params.endDate instanceof Date);
  (0, import_node_assert.default)(params.startDate instanceof Date);
}
function adjustApplicationUri(params) {
  const applicationUri = params.applicationUri || "";
  if (applicationUri.length > 200) {
    throw new Error(`Openssl doesn't support urn with length greater than 200${applicationUri}`);
  }
}

// packages/node-opcua-pki/lib/toolbox/common2.ts
var import_node_assert2 = __toESM(require("assert"));
var import_node_fs = __toESM(require("fs"));
var import_node_path = __toESM(require("path"));
var import_chalk = __toESM(require("chalk"));

// packages/node-opcua-pki/lib/toolbox/config.ts
var g_config = {
  opensslVersion: "unset",
  silent: process.env.VERBOSE ? !process.env.VERBOSE : true,
  force: false
};

// packages/node-opcua-pki/lib/toolbox/debug.ts
var doDebug = process.env.NODEOPCUAPKIDEBUG || false;
var displayError = true;
var displayDebug = !!process.env.NODEOPCUAPKIDEBUG || false;
function debugLog(...args) {
  if (displayDebug) {
    console.log.apply(null, args);
  }
}
function warningLog(...args) {
  console.log.apply(null, args);
}

// packages/node-opcua-pki/lib/toolbox/common2.ts
function certificateFileExist(certificateFile) {
  if (import_node_fs.default.existsSync(certificateFile) && !g_config.force) {
    warningLog(
      import_chalk.default.yellow("        certificate ") + import_chalk.default.cyan(certificateFile) + import_chalk.default.yellow(" already exists => do not overwrite")
    );
    return false;
  }
  return true;
}
function mkdirRecursiveSync(folder) {
  if (!import_node_fs.default.existsSync(folder)) {
    debugLog(import_chalk.default.white(" .. constructing "), folder);
    import_node_fs.default.mkdirSync(folder, { recursive: true });
  }
}
function makePath(folderName, filename) {
  let s;
  if (filename) {
    s = import_node_path.default.join(import_node_path.default.normalize(folderName), filename);
  } else {
    (0, import_node_assert2.default)(folderName);
    s = folderName;
  }
  s = s.replace(/\\/g, "/");
  return s;
}

// packages/node-opcua-pki/lib/toolbox/display.ts
var import_chalk2 = __toESM(require("chalk"));
function displayTitle(str) {
  if (!g_config.silent) {
    warningLog("");
    warningLog(import_chalk2.default.yellowBright(str));
    warningLog(import_chalk2.default.yellow(new Array(str.length + 1).join("=")), "\n");
  }
}
function displaySubtitle(str) {
  if (!g_config.silent) {
    warningLog("");
    warningLog(`    ${import_chalk2.default.yellowBright(str)}`);
    warningLog(`    ${import_chalk2.default.white(new Array(str.length + 1).join("-"))}`, "\n");
  }
}
function display(str) {
  if (!g_config.silent) {
    warningLog(`       ${str}`);
  }
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/index.ts
var import_node_constants = __toESM(require("constants"));

// packages/node-opcua-pki/lib/toolbox/with_openssl/_env.ts
var exportedEnvVars = {};
function setEnv(varName, value) {
  if (!g_config.silent) {
    warningLog(`          set ${varName}=${value}`);
  }
  exportedEnvVars[varName] = value;
  if (["OPENSSL_CONF"].indexOf(varName) >= 0) {
    process.env[varName] = value;
  }
  if (["RANDFILE"].indexOf(varName) >= 0) {
    process.env[varName] = value;
  }
}
function getEnv(varName) {
  return exportedEnvVars[varName];
}
function getEnvironmentVarNames() {
  return Object.keys(exportedEnvVars).map((varName) => {
    return { key: varName, pattern: `\\$ENV\\:\\:${varName}` };
  });
}
function processAltNames(params) {
  params.dns = params.dns || [];
  params.ip = params.ip || [];
  let subjectAltName = [];
  subjectAltName.push(`URI:${params.applicationUri}`);
  subjectAltName = [].concat(
    subjectAltName,
    params.dns.map((d) => `DNS:${d}`)
  );
  subjectAltName = [].concat(
    subjectAltName,
    params.ip.map((d) => `IP:${d}`)
  );
  const subjectAltNameString = subjectAltName.join(", ");
  setEnv("ALTNAME", subjectAltNameString);
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/create_certificate_signing_request.ts
var import_node_assert5 = __toESM(require("assert"));
var import_node_fs5 = __toESM(require("fs"));
var import_node_path4 = __toESM(require("path"));

// packages/node-opcua-pki/lib/misc/subject.ts
var import_node_opcua_crypto = require("node-opcua-crypto");

// packages/node-opcua-pki/lib/toolbox/with_openssl/execute_openssl.ts
var import_node_assert3 = __toESM(require("assert"));
var import_node_child_process2 = __toESM(require("child_process"));
var import_node_fs3 = __toESM(require("fs"));
var import_node_os2 = __toESM(require("os"));
var import_byline2 = __toESM(require("byline"));
var import_chalk4 = __toESM(require("chalk"));

// packages/node-opcua-pki/lib/toolbox/with_openssl/install_prerequisite.ts
var import_node_child_process = __toESM(require("child_process"));
var import_node_fs2 = __toESM(require("fs"));
var import_node_os = __toESM(require("os"));
var import_node_path2 = __toESM(require("path"));
var import_node_url = __toESM(require("url"));
var import_byline = __toESM(require("byline"));
var import_chalk3 = __toESM(require("chalk"));
var import_progress = __toESM(require("progress"));
var import_wget_improved_2 = __toESM(require("wget-improved-2"));
var import_yauzl = __toESM(require("yauzl"));
var doDebug2 = process.env.NODEOPCUAPKIDEBUG || false;
function makeOptions() {
  const proxy = process.env.HTTPS_PROXY || process.env.https_proxy || process.env.HTTP_PROXY || process.env.http_proxy || void 0;
  if (proxy) {
    const a = new import_node_url.default.URL(proxy);
    const auth = a.username ? `${a.username}:${a.password}` : void 0;
    const options = {
      proxy: {
        port: a.port ? parseInt(a.port, 10) : 80,
        protocol: a.protocol.replace(":", ""),
        host: a.hostname ?? "",
        proxyAuth: auth
      }
    };
    warningLog(import_chalk3.default.green("- using proxy "), proxy);
    warningLog(options);
    return options;
  }
  return {};
}
async function execute(cmd, cwd) {
  let output = "";
  const options = {
    cwd,
    windowsHide: true
  };
  return await new Promise((resolve, reject) => {
    const child = import_node_child_process.default.exec(
      cmd,
      options,
      (err) => {
        const exitCode = err === null ? 0 : err.code || 1;
        if (err) reject(err);
        else {
          resolve({ exitCode, output });
        }
      }
    );
    const stream1 = (0, import_byline.default)(child.stdout);
    stream1.on("data", (line) => {
      output += `${line}
`;
      if (doDebug2) {
        process.stdout.write(`        stdout ${import_chalk3.default.yellow(line)}
`);
      }
    });
  });
}
function quote2(str) {
  return `"${str.replace(/\\/g, "/")}"`;
}
function is_expected_openssl_version(strVersion) {
  return !!strVersion.match(/OpenSSL 1|3/);
}
async function getopensslExecPath() {
  let result1;
  try {
    result1 = await execute("which openssl");
  } catch (err) {
    warningLog("warning: ", err.message);
    throw new Error("Cannot find openssl");
  }
  const exitCode = result1?.exitCode;
  const output = result1?.output;
  if (exitCode !== 0) {
    warningLog(import_chalk3.default.yellow(" it seems that ") + import_chalk3.default.cyan("openssl") + import_chalk3.default.yellow(" is not installed on your computer "));
    warningLog(import_chalk3.default.yellow("Please install it before running this programs"));
    throw new Error("Cannot find openssl");
  }
  const opensslExecPath = output.replace(/\n\r/g, "").trim();
  return opensslExecPath;
}
async function check_system_openssl_version() {
  const opensslExecPath = await getopensslExecPath();
  const q_opensslExecPath = quote2(opensslExecPath);
  if (doDebug2) {
    warningLog(`              OpenSSL found in : ${import_chalk3.default.yellow(opensslExecPath)}`);
  }
  const result = await execute(`${q_opensslExecPath} version`);
  const exitCode = result?.exitCode;
  const output = result?.output;
  const version = output.trim();
  const versionOK = exitCode === 0 && is_expected_openssl_version(version);
  if (!versionOK) {
    let message = import_chalk3.default.whiteBright("Warning !!!!!!!!!!!! ") + "\nyour version of openssl is " + version + ". It doesn't match the expected version";
    if (process.platform === "darwin") {
      message += import_chalk3.default.cyan("\nplease refer to :") + import_chalk3.default.yellow(" https://github.com/node-opcua/node-opcua/wiki/installing-node-opcua-or-node-red-on-MacOS");
    }
    console.log(message);
  }
  return output;
}
async function install_and_check_win32_openssl_version() {
  const downloadFolder = import_node_path2.default.join(import_node_os.default.tmpdir(), ".");
  function get_openssl_folder_win32() {
    if (process.env.LOCALAPPDATA) {
      const userProgramFolder = import_node_path2.default.join(process.env.LOCALAPPDATA, "Programs");
      if (import_node_fs2.default.existsSync(userProgramFolder)) {
        return import_node_path2.default.join(userProgramFolder, "openssl");
      }
    }
    return import_node_path2.default.join(process.cwd(), "openssl");
  }
  function get_openssl_exec_path_win32() {
    const opensslFolder2 = get_openssl_folder_win32();
    return import_node_path2.default.join(opensslFolder2, "openssl.exe");
  }
  async function check_openssl_win32() {
    const opensslExecPath2 = get_openssl_exec_path_win32();
    const exists = import_node_fs2.default.existsSync(opensslExecPath2);
    if (!exists) {
      warningLog("checking presence of ", opensslExecPath2);
      warningLog(import_chalk3.default.red(" cannot find file ") + opensslExecPath2);
      return {
        opensslOk: false,
        version: `cannot find file ${opensslExecPath2}`
      };
    } else {
      const q_openssl_exe_path = quote2(opensslExecPath2);
      const cwd = ".";
      const { exitCode, output } = await execute(`${q_openssl_exe_path} version`, cwd);
      const version = output.trim();
      if (doDebug2) {
        warningLog(" Version = ", version);
      }
      return {
        opensslOk: exitCode === 0 && is_expected_openssl_version(version),
        version
      };
    }
  }
  function win32or64() {
    if (process.env.PROCESSOR_ARCHITECTURE === "x86" && process.env.PROCESSOR_ARCHITEW6432) {
      return 64;
    }
    if (process.env.PROCESSOR_ARCHITECTURE === "AMD64") {
      return 64;
    }
    if (process.env.CURRENT_CPU === "x64") {
      return 64;
    }
    return 32;
  }
  async function download_openssl() {
    const url2 = win32or64() === 64 ? "https://github.com/node-opcua/node-opcua-pki/releases/download/2.14.2/openssl-1.0.2u-x64_86-win64.zip" : "https://github.com/node-opcua/node-opcua-pki/releases/download/2.14.2/openssl-1.0.2u-i386-win32.zip";
    const outputFilename = import_node_path2.default.join(downloadFolder, import_node_path2.default.basename(url2));
    warningLog(`downloading ${import_chalk3.default.yellow(url2)} to ${outputFilename}`);
    if (import_node_fs2.default.existsSync(outputFilename)) {
      return { downloadedFile: outputFilename };
    }
    const options = makeOptions();
    const bar = new import_progress.default(import_chalk3.default.cyan("[:bar]") + import_chalk3.default.cyan(" :percent ") + import_chalk3.default.white(":etas"), {
      complete: "=",
      incomplete: " ",
      total: 100,
      width: 100
    });
    return await new Promise((resolve, reject) => {
      const download = import_wget_improved_2.default.download(url2, outputFilename, options);
      download.on("error", (err) => {
        warningLog(err);
        setImmediate(() => {
          reject(err);
        });
      });
      download.on("end", (output) => {
        if (doDebug2) {
          warningLog(output);
        }
        resolve({ downloadedFile: outputFilename });
      });
      download.on("progress", (progress) => {
        bar.update(progress);
      });
    });
  }
  async function unzip_openssl(zipFilename) {
    const opensslFolder2 = get_openssl_folder_win32();
    const zipFile = await new Promise((resolve, reject) => {
      import_yauzl.default.open(zipFilename, { lazyEntries: true }, (err, zipfile) => {
        if (err) {
          reject(err);
        } else {
          if (!zipfile) {
            reject(new Error("zipfile is null"));
          } else {
            resolve(zipfile);
          }
        }
      });
    });
    zipFile.readEntry();
    await new Promise((resolve, reject) => {
      zipFile.on("end", (err) => {
        setImmediate(() => {
          if (doDebug2) {
            warningLog("unzip done");
          }
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        });
      });
      zipFile.on("entry", (entry) => {
        zipFile.openReadStream(entry, (err, readStream) => {
          if (err) {
            return reject(err);
          }
          const file = import_node_path2.default.join(opensslFolder2, entry.fileName);
          if (doDebug2) {
            warningLog(" unzipping :", file);
          }
          const writeStream = import_node_fs2.default.createWriteStream(file, "binary");
          readStream?.pipe(writeStream);
          writeStream.on("close", () => {
            zipFile.readEntry();
          });
        });
      });
    });
  }
  const opensslFolder = get_openssl_folder_win32();
  const opensslExecPath = get_openssl_exec_path_win32();
  if (!import_node_fs2.default.existsSync(opensslFolder)) {
    if (doDebug2) {
      warningLog("creating openssl_folder", opensslFolder);
    }
    import_node_fs2.default.mkdirSync(opensslFolder);
  }
  const { opensslOk, version: _version } = await check_openssl_win32();
  if (!opensslOk) {
    warningLog(import_chalk3.default.yellow("openssl seems to be missing and need to be installed"));
    const { downloadedFile } = await download_openssl();
    if (doDebug2) {
      warningLog("deflating ", import_chalk3.default.yellow(downloadedFile));
    }
    await unzip_openssl(downloadedFile);
    const opensslExists = !!import_node_fs2.default.existsSync(opensslExecPath);
    if (doDebug2) {
      warningLog("verifying ", opensslExists, opensslExists ? import_chalk3.default.green("OK ") : import_chalk3.default.red(" Error"), opensslExecPath);
    }
    const _opensslExecPath2 = await check_openssl_win32();
    return opensslExecPath;
  } else {
    if (doDebug2) {
      warningLog(import_chalk3.default.green("openssl is already installed and have the expected version."));
    }
    return opensslExecPath;
  }
}
async function install_prerequisite() {
  if (process.platform !== "win32") {
    return await check_system_openssl_version();
  } else {
    return await install_and_check_win32_openssl_version();
  }
}
async function get_openssl_exec_path() {
  if (process.platform === "win32") {
    const opensslExecPath = await install_prerequisite();
    if (!import_node_fs2.default.existsSync(opensslExecPath)) {
      throw new Error(`internal error cannot find ${opensslExecPath}`);
    }
    return opensslExecPath;
  } else {
    return "openssl";
  }
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/execute_openssl.ts
var opensslPath;
var n = makePath;
async function execute2(cmd, options) {
  const from = new Error();
  options.cwd = options.cwd || process.cwd();
  if (!g_config.silent) {
    warningLog(import_chalk4.default.cyan("                  CWD         "), options.cwd);
  }
  const outputs = [];
  return await new Promise((resolve, reject) => {
    const child = import_node_child_process2.default.exec(
      cmd,
      {
        cwd: options.cwd,
        windowsHide: true
      },
      (err) => {
        if (err) {
          if (!options.hideErrorMessage) {
            const fence = "###########################################";
            console.error(import_chalk4.default.bgWhiteBright.redBright(`${fence} OPENSSL ERROR ${fence}`));
            console.error(import_chalk4.default.bgWhiteBright.redBright(`CWD = ${options.cwd}`));
            console.error(import_chalk4.default.bgWhiteBright.redBright(err.message));
            console.error(import_chalk4.default.bgWhiteBright.redBright(`${fence} OPENSSL ERROR ${fence}`));
            console.error(from.stack);
          }
          reject(new Error(err.message));
          return;
        }
        resolve(outputs.join(""));
      }
    );
    if (child.stdout) {
      const stream2 = (0, import_byline2.default)(child.stdout);
      stream2.on("data", (line) => {
        outputs.push(`${line}
`);
      });
      if (!g_config.silent) {
        stream2.on("data", (line) => {
          line = line.toString();
          if (doDebug) {
            process.stdout.write(`${import_chalk4.default.white("        stdout ") + import_chalk4.default.whiteBright(line)}
`);
          }
        });
      }
    }
    if (!g_config.silent) {
      if (child.stderr) {
        const stream1 = (0, import_byline2.default)(child.stderr);
        stream1.on("data", (line) => {
          line = line.toString();
          if (displayError) {
            process.stdout.write(`${import_chalk4.default.white("        stderr ") + import_chalk4.default.red(line)}
`);
          }
        });
      }
    }
  });
}
async function find_openssl() {
  return await get_openssl_exec_path();
}
async function ensure_openssl_installed() {
  if (!opensslPath) {
    opensslPath = await find_openssl();
    const outputs = await execute_openssl("version", { cwd: "." });
    g_config.opensslVersion = outputs.trim();
    if (doDebug) {
      warningLog("OpenSSL version : ", g_config.opensslVersion);
    }
  }
}
async function execute_openssl_no_failure(cmd, options) {
  options = options || {};
  options.hideErrorMessage = true;
  try {
    return await execute_openssl(cmd, options);
  } catch (err) {
    debugLog(" (ignored error =  ERROR : )", err.message);
  }
}
function getTempFolder() {
  return import_node_os2.default.tmpdir();
}
async function execute_openssl(cmd, options) {
  debugLog("execute_openssl", cmd, options);
  const empty_config_file = n(getTempFolder(), "empty_config.cnf");
  if (!import_node_fs3.default.existsSync(empty_config_file)) {
    await import_node_fs3.default.promises.writeFile(empty_config_file, "# empty config file");
  }
  options = options || {};
  options.openssl_conf = options.openssl_conf || empty_config_file;
  (0, import_node_assert3.default)(options.openssl_conf);
  setEnv("OPENSSL_CONF", options.openssl_conf);
  if (!g_config.silent) {
    warningLog(import_chalk4.default.cyan("                  OPENSSL_CONF"), process.env.OPENSSL_CONF);
    warningLog(import_chalk4.default.cyan("                  RANDFILE    "), process.env.RANDFILE);
    warningLog(import_chalk4.default.cyan("                  CMD         openssl "), import_chalk4.default.cyanBright(cmd));
  }
  await ensure_openssl_installed();
  return await execute2(`${quote(opensslPath)} ${cmd}`, options);
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/toolbox.ts
var import_node_assert4 = __toESM(require("assert"));
var import_node_fs4 = __toESM(require("fs"));
var import_node_path3 = __toESM(require("path"));
function openssl_require2DigitYearInDate() {
  if (!g_config.opensslVersion) {
    throw new Error(
      "openssl_require2DigitYearInDate : openssl version is not known:  please call ensure_openssl_installed()"
    );
  }
  return g_config.opensslVersion.match(/OpenSSL 0\.9/);
}
g_config.opensslVersion = "";
var _counter = 0;
function generateStaticConfig(configPath, options) {
  const prePath = options?.cwd || "";
  const originalFilename = !import_node_path3.default.isAbsolute(configPath) ? import_node_path3.default.join(prePath, configPath) : configPath;
  let staticConfig = import_node_fs4.default.readFileSync(originalFilename, { encoding: "utf8" });
  for (const envVar of getEnvironmentVarNames()) {
    staticConfig = staticConfig.replace(new RegExp(envVar.pattern, "gi"), getEnv(envVar.key));
  }
  const staticConfigPath = `${configPath}.${process.pid}-${_counter++}.tmp`;
  const temporaryConfigPath = !import_node_path3.default.isAbsolute(configPath) ? import_node_path3.default.join(prePath, staticConfigPath) : staticConfigPath;
  import_node_fs4.default.writeFileSync(temporaryConfigPath, staticConfig);
  if (options?.cwd) {
    return import_node_path3.default.relative(options.cwd, temporaryConfigPath);
  } else {
    return temporaryConfigPath;
  }
}
function x509Date(date) {
  date = date || /* @__PURE__ */ new Date();
  const Y = date.getUTCFullYear();
  const M = date.getUTCMonth() + 1;
  const D = date.getUTCDate();
  const h = date.getUTCHours();
  const m = date.getUTCMinutes();
  const s = date.getUTCSeconds();
  function w(s2, l) {
    return `${s2}`.padStart(l, "0");
  }
  if (openssl_require2DigitYearInDate()) {
    return `${w(Y, 2) + w(M, 2) + w(D, 2) + w(h, 2) + w(m, 2) + w(s, 2)}Z`;
  } else {
    return `${w(Y, 4) + w(M, 2) + w(D, 2) + w(h, 2) + w(m, 2) + w(s, 2)}Z`;
  }
}

// packages/node-opcua-pki/lib/ca/templates/ca_config_template.cnf.ts
var config = `#.........DO NOT MODIFY BY HAND .........................
[ ca ]
default_ca               = CA_default
[ CA_default ]
dir                      = %%ROOT_FOLDER%%            # the main CA folder
certs                    = $dir/certs                 # where to store certificates
new_certs_dir            = $dir/certs                 #
database                 = $dir/index.txt             # the certificate database
serial                   = $dir/serial                # the serial number counter
certificate              = $dir/public/cacert.pem     # The root CA certificate
private_key              = $dir/private/cakey.pem     # the CA private key
x509_extensions          = usr_cert                   #
default_days             = 3650                       # default validity : 10 years

# default_md               = sha1

default_md                = sha256                      # The default digest algorithm

preserve                 = no
policy                   = policy_match
# randfile                 = $dir/random.rnd
# default_startdate        = YYMMDDHHMMSSZ
# default_enddate          = YYMMDDHHMMSSZ
crl_dir                  = $dir/crl
crl_extensions           = crl_ext
crl                      = $dir/revocation_list.crl # the Revocation list
crlnumber                = $dir/crlnumber           # CRL number file
default_crl_days         = 30
default_crl_hours        = 24
#msie_hack

[ policy_match ]
countryName              = optional
stateOrProvinceName      = optional
localityName             = optional
organizationName         = optional
organizationalUnitName   = optional
commonName               = optional
emailAddress             = optional

[ req ]
default_bits             = 4096                     # Size of keys
default_keyfile          = key.pem                  # name of generated keys
distinguished_name       = req_distinguished_name
attributes               = req_attributes
x509_extensions          = v3_ca
#input_password
#output_password
string_mask              = nombstr                  # permitted characters
req_extensions           = v3_req

[ req_distinguished_name ]

#0 countryName             = Country Name (2 letter code)
# countryName_default     = FR
# countryName_min         = 2
# countryName_max         = 2
# stateOrProvinceName     = State or Province Name (full name)
# stateOrProvinceName_default = Ile de France
# localityName            = Locality Name (city, district)
# localityName_default    = Paris
organizationName          = Organization Name (company)
organizationName_default  = NodeOPCUA
# organizationalUnitName  = Organizational Unit Name (department, division)
# organizationalUnitName_default = R&D
commonName                = Common Name (hostname, FQDN, IP, or your name)
commonName_max            = 256
commonName_default        = NodeOPCUA
# emailAddress            = Email Address
# emailAddress_max        = 40
# emailAddress_default    = node-opcua (at) node-opcua (dot) com

[ req_attributes ]
#challengePassword        = A challenge password
#challengePassword_min    = 4
#challengePassword_max    = 20
#unstructuredName         = An optional company name
[ usr_cert ]
basicConstraints          = critical, CA:FALSE
subjectKeyIdentifier      = hash
authorityKeyIdentifier    = keyid,issuer:always
#authorityKeyIdentifier    = keyid
subjectAltName            = $ENV::ALTNAME
# issuerAltName            = issuer:copy
nsComment                 = ''OpenSSL Generated Certificate''
#nsCertType               = client, email, objsign for ''everything including object signing''
#nsCaRevocationUrl        = http://www.domain.dom/ca-crl.pem
#nsBaseUrl                =
#nsRenewalUrl             =
#nsCaPolicyUrl            =
#nsSslServerName          =
keyUsage                  = critical, digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment, keyAgreement, keyCertSign
extendedKeyUsage          = critical,serverAuth ,clientAuth

[ v3_req ]
basicConstraints          = critical, CA:FALSE
keyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyAgreement
extendedKeyUsage          = critical,serverAuth ,clientAuth
subjectAltName            = $ENV::ALTNAME
nsComment                 = "CA Generated by Node-OPCUA Certificate utility using openssl"
[ v3_ca ]
subjectKeyIdentifier      = hash
authorityKeyIdentifier    = keyid:always,issuer:always
# authorityKeyIdentifier    = keyid
basicConstraints          = CA:TRUE
keyUsage                  = critical, cRLSign, keyCertSign
nsComment                 = "CA Certificate generated by Node-OPCUA Certificate utility using openssl"
#nsCertType                 = sslCA, emailCA
#subjectAltName             = email:copy
#issuerAltName              = issuer:copy
#obj                        = DER:02:03
crlDistributionPoints     = @crl_info
[ crl_info ]
URI.0                     = http://localhost:8900/crl.pem
[ v3_selfsigned]
basicConstraints          = critical, CA:FALSE
keyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyAgreement
extendedKeyUsage          = critical,serverAuth ,clientAuth
nsComment                 = "Self-signed certificate, generated by NodeOPCUA"
subjectAltName            = $ENV::ALTNAME

[ crl_ext ]
#issuerAltName            = issuer:copy
authorityKeyIdentifier    = keyid:always,issuer:always
#authorityInfoAccess       = @issuer_info`;
var ca_config_template_cnf_default = config;

// packages/node-opcua-pki/lib/ca/certificate_authority.ts
var defaultSubject = "/C=FR/ST=IDF/L=Paris/O=Local NODE-OPCUA Certificate Authority/CN=NodeOPCUA-CA";
var configurationFileTemplate = ca_config_template_cnf_default;
var config2 = {
  certificateDir: "INVALID",
  forceCA: false,
  pkiDir: "INVALID"
};
var n2 = makePath;
var q = quote;
function octetStringToIpAddress(a) {
  return parseInt(a.substring(0, 2), 16).toString() + "." + parseInt(a.substring(2, 4), 16).toString() + "." + parseInt(a.substring(4, 6), 16).toString() + "." + parseInt(a.substring(6, 8), 16).toString();
}
(0, import_node_assert6.default)(octetStringToIpAddress("c07b9179") === "192.123.145.121");
async function construct_CertificateAuthority(certificateAuthority) {
  const subject = certificateAuthority.subject;
  const caRootDir = import_node_path5.default.resolve(certificateAuthority.rootDir);
  async function make_folders() {
    mkdirRecursiveSync(caRootDir);
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "private"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "public"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "certs"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "crl"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "conf"));
  }
  await make_folders();
  async function construct_default_files() {
    const serial = import_node_path5.default.join(caRootDir, "serial");
    if (!import_node_fs6.default.existsSync(serial)) {
      await import_node_fs6.default.promises.writeFile(serial, "1000");
    }
    const crlNumber = import_node_path5.default.join(caRootDir, "crlnumber");
    if (!import_node_fs6.default.existsSync(crlNumber)) {
      await import_node_fs6.default.promises.writeFile(crlNumber, "1000");
    }
    const indexFile = import_node_path5.default.join(caRootDir, "index.txt");
    if (!import_node_fs6.default.existsSync(indexFile)) {
      await import_node_fs6.default.promises.writeFile(indexFile, "");
    }
  }
  await construct_default_files();
  if (import_node_fs6.default.existsSync(import_node_path5.default.join(caRootDir, "private/cakey.pem")) && !config2.forceCA) {
    debugLog("CA private key already exists ... skipping");
    return;
  }
  displayTitle("Create Certificate Authority (CA)");
  const indexFileAttr = import_node_path5.default.join(caRootDir, "index.txt.attr");
  if (!import_node_fs6.default.existsSync(indexFileAttr)) {
    await import_node_fs6.default.promises.writeFile(indexFileAttr, "unique_subject = no");
  }
  const caConfigFile = certificateAuthority.configFile;
  if (1) {
    let data = configurationFileTemplate;
    data = makePath(data.replace(/%%ROOT_FOLDER%%/, caRootDir));
    await import_node_fs6.default.promises.writeFile(caConfigFile, data);
  }
  const subjectOpt = ` -subj "${subject.toString()}" `;
  processAltNames({});
  const options = { cwd: caRootDir };
  const configFile = generateStaticConfig("conf/caconfig.cnf", options);
  const configOption = ` -config ${q(n2(configFile))}`;
  const keySize = certificateAuthority.keySize;
  const privateKeyFilename = import_node_path5.default.join(caRootDir, "private/cakey.pem");
  const csrFilename = import_node_path5.default.join(caRootDir, "private/cakey.csr");
  displayTitle(`Generate the CA private Key - ${keySize}`);
  await (0, import_node_opcua_crypto2.generatePrivateKeyFile)(privateKeyFilename, keySize);
  displayTitle("Generate a certificate request for the CA key");
  await execute_openssl(
    "req -new -sha256  -text  -extensions v3_ca" + configOption + " -key " + q(n2(privateKeyFilename)) + " -out " + q(n2(csrFilename)) + " " + subjectOpt,
    options
  );
  displayTitle("Generate CA Certificate (self-signed)");
  await execute_openssl(
    " x509 -sha256 -req -days 3650  -text  -extensions v3_ca -extfile " + q(n2(configFile)) + " -in private/cakey.csr  -signkey " + q(n2(privateKeyFilename)) + " -out public/cacert.pem",
    options
  );
  displaySubtitle("generate initial CRL (Certificate Revocation List)");
  await regenerateCrl(certificateAuthority.revocationList, configOption, options);
  displayTitle("Create Certificate Authority (CA) ---> DONE");
}
async function regenerateCrl(revocationList, configOption, options) {
  displaySubtitle("regenerate CRL (Certificate Revocation List)");
  await execute_openssl(`ca -gencrl ${configOption} -out crl/revocation_list.crl`, options);
  await execute_openssl("crl  -in  crl/revocation_list.crl -out  crl/revocation_list.der  -outform der", options);
  displaySubtitle("Display (Certificate Revocation List)");
  await execute_openssl(`crl  -in ${q(n2(revocationList))} -text  -noout`, options);
}
var CertificateAuthority = class {
  /** RSA key size used when generating the CA private key. */
  keySize;
  /** Root filesystem path of the CA directory structure. */
  location;
  /** X.500 subject of the CA certificate. */
  subject;
  constructor(options) {
    (0, import_node_assert6.default)(Object.prototype.hasOwnProperty.call(options, "location"));
    (0, import_node_assert6.default)(Object.prototype.hasOwnProperty.call(options, "keySize"));
    this.location = options.location;
    this.keySize = options.keySize || 2048;
    this.subject = new import_node_opcua_crypto2.Subject(options.subject || defaultSubject);
  }
  /** Absolute path to the CA root directory (alias for {@link location}). */
  get rootDir() {
    return this.location;
  }
  /** Path to the OpenSSL configuration file (`conf/caconfig.cnf`). */
  get configFile() {
    return import_node_path5.default.normalize(import_node_path5.default.join(this.rootDir, "./conf/caconfig.cnf"));
  }
  /** Path to the CA certificate in PEM format (`public/cacert.pem`). */
  get caCertificate() {
    return makePath(this.rootDir, "./public/cacert.pem");
  }
  /**
   * Path to the current Certificate Revocation List in DER format.
   * (`crl/revocation_list.der`)
   */
  get revocationListDER() {
    return makePath(this.rootDir, "./crl/revocation_list.der");
  }
  /**
   * Path to the current Certificate Revocation List in PEM format.
   * (`crl/revocation_list.crl`)
   */
  get revocationList() {
    return makePath(this.rootDir, "./crl/revocation_list.crl");
  }
  /**
   * Path to the concatenated CA certificate + CRL file.
   * Used by OpenSSL for CRL-based verification.
   */
  get caCertificateWithCrl() {
    return makePath(this.rootDir, "./public/cacertificate_with_crl.pem");
  }
  /**
   * Initialize the CA directory structure, generate the CA
   * private key and self-signed certificate if they do not
   * already exist.
   */
  async initialize() {
    await construct_CertificateAuthority(this);
  }
  /**
   * Rebuild the combined CA certificate + CRL file.
   *
   * This concatenates the CA certificate with the current
   * revocation list so that OpenSSL can verify certificates
   * with CRL checking enabled.
   */
  async constructCACertificateWithCRL() {
    const cacertWithCRL = this.caCertificateWithCrl;
    if (import_node_fs6.default.existsSync(this.revocationList)) {
      await import_node_fs6.default.promises.writeFile(
        cacertWithCRL,
        import_node_fs6.default.readFileSync(this.caCertificate, "utf8") + import_node_fs6.default.readFileSync(this.revocationList, "utf8")
      );
    } else {
      await import_node_fs6.default.promises.writeFile(cacertWithCRL, import_node_fs6.default.readFileSync(this.caCertificate));
    }
  }
  /**
   * Append the CA certificate to a signed certificate file,
   * creating a PEM certificate chain.
   *
   * @param certificate - path to the certificate file to extend
   */
  async constructCertificateChain(certificate) {
    (0, import_node_assert6.default)(import_node_fs6.default.existsSync(certificate));
    (0, import_node_assert6.default)(import_node_fs6.default.existsSync(this.caCertificate));
    debugLog(import_chalk5.default.yellow("        certificate file :"), import_chalk5.default.cyan(certificate));
    await import_node_fs6.default.promises.writeFile(
      certificate,
      await import_node_fs6.default.promises.readFile(certificate, "utf8") + await import_node_fs6.default.promises.readFile(this.caCertificate, "utf8")
      //   + fs.readFileSync(this.revocationList)
    );
  }
  /**
   * Create a self-signed certificate using OpenSSL.
   *
   * @param certificateFile - output path for the signed certificate
   * @param privateKey - path to the private key file
   * @param params - certificate parameters (subject, validity, SANs)
   */
  async createSelfSignedCertificate(certificateFile, privateKey, params) {
    (0, import_node_assert6.default)(typeof privateKey === "string");
    (0, import_node_assert6.default)(import_node_fs6.default.existsSync(privateKey));
    if (!certificateFileExist(certificateFile)) {
      return;
    }
    adjustDate(params);
    adjustApplicationUri(params);
    processAltNames(params);
    const csrFile = `${certificateFile}_csr`;
    (0, import_node_assert6.default)(csrFile);
    const configFile = generateStaticConfig(this.configFile, { cwd: this.rootDir });
    const options = {
      cwd: this.rootDir,
      openssl_conf: makePath(configFile)
    };
    const configOption = "";
    const subject = params.subject ? new import_node_opcua_crypto2.Subject(params.subject).toString() : "";
    const subjectOptions = subject && subject.length > 1 ? ` -subj ${subject} ` : "";
    displaySubtitle("- the certificate signing request");
    await execute_openssl(
      "req  -new -sha256 -text " + configOption + subjectOptions + " -batch -key " + q(n2(privateKey)) + " -out " + q(n2(csrFile)),
      options
    );
    displaySubtitle("- creating the self-signed certificate");
    await execute_openssl(
      "ca  -selfsign  -keyfile " + q(n2(privateKey)) + " -startdate " + x509Date(params.startDate) + " -enddate " + x509Date(params.endDate) + " -batch -out " + q(n2(certificateFile)) + " -in " + q(n2(csrFile)),
      options
    );
    displaySubtitle("- dump the certificate for a check");
    await execute_openssl(`x509 -in ${q(n2(certificateFile))}  -dates -fingerprint -purpose -noout`, {});
    displaySubtitle("- verify self-signed certificate");
    await execute_openssl_no_failure(`verify -verbose -CAfile ${q(n2(certificateFile))} ${q(n2(certificateFile))}`, options);
    await import_node_fs6.default.promises.unlink(csrFile);
  }
  /**
   * Revoke a certificate and regenerate the CRL.
   *
   * @param certificate - path to the certificate file to revoke
   * @param params - revocation parameters
   * @param params.reason - CRL reason code
   *   (default `"keyCompromise"`)
   */
  async revokeCertificate(certificate, params) {
    const crlReasons = [
      "unspecified",
      "keyCompromise",
      "CACompromise",
      "affiliationChanged",
      "superseded",
      "cessationOfOperation",
      "certificateHold",
      "removeFromCRL"
    ];
    const configFile = generateStaticConfig("conf/caconfig.cnf", { cwd: this.rootDir });
    const options = {
      cwd: this.rootDir,
      openssl_conf: makePath(configFile)
    };
    setEnv("ALTNAME", "");
    const randomFile = import_node_path5.default.join(this.rootDir, "random.rnd");
    setEnv("RANDFILE", randomFile);
    const configOption = ` -config ${q(n2(configFile))}`;
    const reason = params.reason || "keyCompromise";
    (0, import_node_assert6.default)(crlReasons.indexOf(reason) >= 0);
    displayTitle(`Revoking certificate  ${certificate}`);
    displaySubtitle("Revoke certificate");
    await execute_openssl_no_failure(`ca -verbose ${configOption} -revoke ${q(certificate)} -crl_reason ${reason}`, options);
    await regenerateCrl(this.revocationList, configOption, options);
    displaySubtitle("Verify that certificate is revoked");
    await execute_openssl_no_failure(
      "verify -verbose -CRLfile " + q(n2(this.revocationList)) + " -CAfile " + q(n2(this.caCertificate)) + " -crl_check " + q(n2(certificate)),
      options
    );
    displaySubtitle("Produce CRL in DER form ");
    await execute_openssl(`crl  -in ${q(n2(this.revocationList))} -out crl/revocation_list.der  -outform der`, options);
    displaySubtitle("Produce CRL in PEM form ");
    await execute_openssl(`crl  -in ${q(n2(this.revocationList))} -out crl/revocation_list.pem  -outform pem -text `, options);
  }
  /**
   * Sign a Certificate Signing Request (CSR) with this CA.
   *
   * The signed certificate is written to `certificate`, and the
   * CA certificate chain plus CRL are appended to form a
   * complete certificate chain.
   *
   * @param certificate - output path for the signed certificate
   * @param certificateSigningRequestFilename - path to the CSR
   * @param params1 - signing parameters (validity, dates, SANs)
   * @returns the path to the signed certificate
   */
  async signCertificateRequest(certificate, certificateSigningRequestFilename, params1) {
    await ensure_openssl_installed();
    (0, import_node_assert6.default)(import_node_fs6.default.existsSync(certificateSigningRequestFilename));
    if (!certificateFileExist(certificate)) {
      return "";
    }
    adjustDate(params1);
    adjustApplicationUri(params1);
    processAltNames(params1);
    const options = { cwd: this.rootDir };
    const csr = await (0, import_node_opcua_crypto2.readCertificateSigningRequest)(certificateSigningRequestFilename);
    const csrInfo = (0, import_node_opcua_crypto2.exploreCertificateSigningRequest)(csr);
    const applicationUri = csrInfo.extensionRequest.subjectAltName.uniformResourceIdentifier ? csrInfo.extensionRequest.subjectAltName.uniformResourceIdentifier[0] : void 0;
    if (typeof applicationUri !== "string") {
      throw new Error("Cannot find applicationUri in CSR");
    }
    const dns = csrInfo.extensionRequest.subjectAltName.dNSName || [];
    let ip = csrInfo.extensionRequest.subjectAltName.iPAddress || [];
    ip = ip.map(octetStringToIpAddress);
    const params = {
      applicationUri,
      dns,
      ip
    };
    processAltNames(params);
    const configFile = generateStaticConfig("conf/caconfig.cnf", options);
    displaySubtitle("- then we ask the authority to sign the certificate signing request");
    const configOption = ` -config ${configFile}`;
    await execute_openssl(
      "ca " + configOption + " -startdate " + x509Date(params1.startDate) + " -enddate " + x509Date(params1.endDate) + " -batch -out " + q(n2(certificate)) + " -in " + q(n2(certificateSigningRequestFilename)),
      options
    );
    displaySubtitle("- dump the certificate for a check");
    await execute_openssl(`x509 -in ${q(n2(certificate))}  -dates -fingerprint -purpose -noout`, options);
    displaySubtitle("- construct CA certificate with CRL");
    await this.constructCACertificateWithCRL();
    displaySubtitle("- construct certificate chain");
    await this.constructCertificateChain(certificate);
    displaySubtitle("- verify certificate against the root CA");
    await this.verifyCertificate(certificate);
    return certificate;
  }
  /**
   * Verify a certificate against this CA.
   *
   * @param certificate - path to the certificate file to verify
   */
  async verifyCertificate(certificate) {
    const isImplemented = false;
    if (isImplemented) {
      const options = { cwd: this.rootDir };
      const configFile = generateStaticConfig("conf/caconfig.cnf", options);
      setEnv("OPENSSL_CONF", makePath(configFile));
      const _configOption = ` -config ${configFile}`;
      _configOption;
      await execute_openssl_no_failure(
        `verify -verbose  -CAfile ${q(n2(this.caCertificateWithCrl))} ${q(n2(certificate))}`,
        options
      );
    }
  }
};

// packages/node-opcua-pki/lib/pki/certificate_manager.ts
var import_node_events = require("events");
var import_node_fs9 = __toESM(require("fs"));
var import_node_path6 = __toESM(require("path"));
var import_global_mutex = require("@ster5/global-mutex");
var import_chalk6 = __toESM(require("chalk"));
var import_chokidar = __toESM(require("chokidar"));
var import_node_opcua_crypto5 = require("node-opcua-crypto");

// packages/node-opcua-pki/lib/toolbox/without_openssl/create_certificate_signing_request.ts
var import_node_assert7 = __toESM(require("assert"));
var import_node_fs7 = __toESM(require("fs"));
var import_node_opcua_crypto3 = require("node-opcua-crypto");
async function createCertificateSigningRequestAsync(certificateSigningRequestFilename, params) {
  (0, import_node_assert7.default)(params);
  (0, import_node_assert7.default)(params.rootDir);
  (0, import_node_assert7.default)(params.configFile);
  (0, import_node_assert7.default)(params.privateKey);
  (0, import_node_assert7.default)(typeof params.privateKey === "string");
  (0, import_node_assert7.default)(import_node_fs7.default.existsSync(params.privateKey), `Private key must exist${params.privateKey}`);
  (0, import_node_assert7.default)(import_node_fs7.default.existsSync(params.rootDir), "RootDir key must exist");
  (0, import_node_assert7.default)(typeof certificateSigningRequestFilename === "string");
  const subject = params.subject ? new import_node_opcua_crypto3.Subject(params.subject).toString() : void 0;
  displaySubtitle("- Creating a Certificate Signing Request with subtile");
  const privateKeyPem = await import_node_fs7.default.promises.readFile(params.privateKey, "utf-8");
  const privateKey = await (0, import_node_opcua_crypto3.pemToPrivateKey)(privateKeyPem);
  const { csr } = await (0, import_node_opcua_crypto3.createCertificateSigningRequest)({
    privateKey,
    dns: params.dns,
    ip: params.ip,
    subject,
    applicationUri: params.applicationUri,
    purpose: params.purpose
  });
  await import_node_fs7.default.promises.writeFile(certificateSigningRequestFilename, csr, "utf-8");
  display(`- privateKey ${params.privateKey}`);
  display(`- certificateSigningRequestFilename ${certificateSigningRequestFilename}`);
}

// packages/node-opcua-pki/lib/toolbox/without_openssl/create_self_signed_certificate.ts
var import_node_assert8 = __toESM(require("assert"));
var import_node_fs8 = __toESM(require("fs"));
var import_node_opcua_crypto4 = require("node-opcua-crypto");
async function createSelfSignedCertificateAsync(certificate, params) {
  params.purpose = params.purpose || import_node_opcua_crypto4.CertificatePurpose.ForApplication;
  (0, import_node_assert8.default)(params.purpose, "Please provide a Certificate Purpose");
  (0, import_node_assert8.default)(import_node_fs8.default.existsSync(params.configFile));
  (0, import_node_assert8.default)(import_node_fs8.default.existsSync(params.rootDir));
  (0, import_node_assert8.default)(import_node_fs8.default.existsSync(params.privateKey));
  if (!params.subject) {
    throw Error("Missing subject");
  }
  (0, import_node_assert8.default)(typeof params.applicationUri === "string");
  (0, import_node_assert8.default)(Array.isArray(params.dns));
  adjustDate(params);
  (0, import_node_assert8.default)(Object.prototype.hasOwnProperty.call(params, "validity"));
  let subject = new import_node_opcua_crypto4.Subject(params.subject);
  subject = subject.toString();
  const purpose = params.purpose;
  displayTitle("Generate a certificate request");
  const privateKeyPem = await import_node_fs8.default.promises.readFile(params.privateKey, "utf-8");
  const privateKey = await (0, import_node_opcua_crypto4.pemToPrivateKey)(privateKeyPem);
  const { cert } = await (0, import_node_opcua_crypto4.createSelfSignedCertificate)({
    privateKey,
    notBefore: params.startDate,
    notAfter: params.endDate,
    validity: params.validity,
    dns: params.dns,
    ip: params.ip,
    subject,
    applicationUri: params.applicationUri,
    purpose
  });
  await import_node_fs8.default.promises.writeFile(certificate, cert, "utf-8");
}
async function createSelfSignedCertificate(certificate, params) {
  await createSelfSignedCertificateAsync(certificate, params);
}

// packages/node-opcua-pki/lib/pki/templates/simple_config_template.cnf.ts
var config3 = '##################################################################################################\n## SIMPLE OPENSSL CONFIG FILE FOR SELF-SIGNED CERTIFICATE GENERATION\n################################################################################################################\n\ndistinguished_name       = req_distinguished_name\ndefault_md               = sha1\n\ndefault_md                = sha256                      # The default digest algorithm\n\n[ v3_ca ]\nsubjectKeyIdentifier        = hash\nauthorityKeyIdentifier      = keyid:always,issuer:always\n\n# authorityKeyIdentifier    = keyid\nbasicConstraints            = CA:TRUE\nkeyUsage                    = critical, cRLSign, keyCertSign\nnsComment                   = "Self-signed Certificate for CA generated by Node-OPCUA Certificate utility"\n#nsCertType                 = sslCA, emailCA\n#subjectAltName             = email:copy\n#issuerAltName              = issuer:copy\n#obj                        = DER:02:03\n# crlDistributionPoints       = @crl_info\n# [ crl_info ]\n# URI.0                     = http://localhost:8900/crl.pem\nsubjectAltName              = $ENV::ALTNAME\n\n[ req ]\ndays                      = 390\nreq_extensions            = v3_req\nx509_extensions           = v3_ca\n\n[v3_req]\nbasicConstraints       = CA:false\nkeyUsage               = critical, cRLSign, keyCertSign\nsubjectAltName         = $ENV::ALTNAME\n\n[ v3_ca_signed]\nsubjectKeyIdentifier      = hash\nauthorityKeyIdentifier    = keyid,issuer\nbasicConstraints          = critical, CA:FALSE\nkeyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyCertSign\nextendedKeyUsage          = clientAuth,serverAuth \nnsComment                 = "certificate generated by Node-OPCUA Certificate utility and signed by a CA"\nsubjectAltName            = $ENV::ALTNAME\n[ v3_selfsigned]\nsubjectKeyIdentifier      = hash\nauthorityKeyIdentifier    = keyid,issuer\nbasicConstraints          = critical, CA:FALSE\nkeyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyCertSign\nextendedKeyUsage          = clientAuth,serverAuth \nnsComment                 = "Self-signed certificate generated by Node-OPCUA Certificate utility"\nsubjectAltName            = $ENV::ALTNAME\n[ req_distinguished_name ]\ncountryName             = Country Name (2 letter code)\ncountryName_default     = FR\ncountryName_min         = 2\ncountryName_max         = 2\n# stateOrProvinceName     = State or Province Name (full name)\n# stateOrProvinceName_default = Ile de France\n# localityName            = Locality Name (city, district)\n# localityName_default    = Paris\norganizationName          = Organization Name (company)\norganizationName_default  = NodeOPCUA\n# organizationalUnitName  = Organizational Unit Name (department, division)\n# organizationalUnitName_default = R&D\ncommonName                = Common Name (hostname, FQDN, IP, or your name)\ncommonName_max            = 256\ncommonName_default        = NodeOPCUA\n# emailAddress            = Email Address\n# emailAddress_max        = 40\n# emailAddress_default    = node-opcua (at) node-opcua (dot) com\nsubjectAltName            = $ENV::ALTNAME';
var simple_config_template_cnf_default = config3;

// packages/node-opcua-pki/lib/pki/certificate_manager.ts
var configurationFileSimpleTemplate = simple_config_template_cnf_default;
var fsWriteFile = import_node_fs9.default.promises.writeFile;
function getOrComputeInfo(entry) {
  if (!entry.info) {
    entry.info = exploreCertificateCached(entry.certificate);
  }
  return entry.info;
}
var EXPLORE_CACHE_MAX = 8;
var _exploreCache = /* @__PURE__ */ new Map();
function exploreCertificateCached(certificate) {
  const key = (0, import_node_opcua_crypto5.makeSHA1Thumbprint)(certificate).toString("hex");
  const cached = _exploreCache.get(key);
  if (cached) {
    _exploreCache.delete(key);
    _exploreCache.set(key, cached);
    return cached;
  }
  const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
  _exploreCache.set(key, info);
  if (_exploreCache.size > EXPLORE_CACHE_MAX) {
    const oldest = _exploreCache.keys().next().value;
    if (oldest) _exploreCache.delete(oldest);
  }
  return info;
}
var VerificationStatus = /* @__PURE__ */ ((VerificationStatus2) => {
  VerificationStatus2["BadCertificateInvalid"] = "BadCertificateInvalid";
  VerificationStatus2["BadSecurityChecksFailed"] = "BadSecurityChecksFailed";
  VerificationStatus2["BadCertificatePolicyCheckFailed"] = "BadCertificatePolicyCheckFailed";
  VerificationStatus2["BadCertificateTimeInvalid"] = "BadCertificateTimeInvalid";
  VerificationStatus2["BadCertificateIssuerTimeInvalid"] = "BadCertificateIssuerTimeInvalid";
  VerificationStatus2["BadCertificateHostNameInvalid"] = "BadCertificateHostNameInvalid";
  VerificationStatus2["BadCertificateUriInvalid"] = "BadCertificateUriInvalid";
  VerificationStatus2["BadCertificateUseNotAllowed"] = "BadCertificateUseNotAllowed";
  VerificationStatus2["BadCertificateIssuerUseNotAllowed"] = "BadCertificateIssuerUseNotAllowed";
  VerificationStatus2["BadCertificateUntrusted"] = "BadCertificateUntrusted";
  VerificationStatus2["BadCertificateRevocationUnknown"] = "BadCertificateRevocationUnknown";
  VerificationStatus2["BadCertificateIssuerRevocationUnknown"] = "BadCertificateIssuerRevocationUnknown";
  VerificationStatus2["BadCertificateRevoked"] = "BadCertificateRevoked";
  VerificationStatus2["BadCertificateIssuerRevoked"] = "BadCertificateIssuerRevoked";
  VerificationStatus2["BadCertificateChainIncomplete"] = "BadCertificateChainIncomplete";
  VerificationStatus2["Good"] = "Good";
  return VerificationStatus2;
})(VerificationStatus || {});
function makeFingerprint(certificate) {
  const chain = (0, import_node_opcua_crypto5.split_der)(certificate);
  return (0, import_node_opcua_crypto5.makeSHA1Thumbprint)(chain[0]).toString("hex");
}
function short(stringToShorten) {
  return stringToShorten.substring(0, 10);
}
var forbiddenChars = /[\x00-\x1F<>:"/\\|?*]/g;
function buildIdealCertificateName(certificate) {
  const fingerprint = makeFingerprint(certificate);
  try {
    const commonName = exploreCertificateCached(certificate).tbsCertificate.subject.commonName || "";
    const sanitizedCommonName = commonName.replace(forbiddenChars, "_");
    return `${sanitizedCommonName}[${fingerprint}]`;
  } catch (_err) {
    return `invalid_certificate_[${fingerprint}]`;
  }
}
function findMatchingIssuerKey(entries, wantedIssuerKey) {
  return entries.filter((entry) => {
    const info = getOrComputeInfo(entry);
    return info.tbsCertificate.extensions && info.tbsCertificate.extensions.subjectKeyIdentifier === wantedIssuerKey;
  });
}
function isSelfSigned2(info) {
  return info.tbsCertificate.extensions?.subjectKeyIdentifier === info.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
}
function isSelfSigned3(certificate) {
  const info = exploreCertificateCached(certificate);
  return isSelfSigned2(info);
}
function findIssuerCertificateInChain(certificate, chain) {
  if (!certificate) {
    return null;
  }
  const certInfo = exploreCertificateCached(certificate);
  if (isSelfSigned2(certInfo)) {
    return certificate;
  }
  const wantedIssuerKey = certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
  if (!wantedIssuerKey) {
    debugLog("Certificate has no extension 3");
    return null;
  }
  const potentialIssuers = chain.filter((c) => {
    const info = exploreCertificateCached(c);
    return info.tbsCertificate.extensions && info.tbsCertificate.extensions.subjectKeyIdentifier === wantedIssuerKey;
  });
  if (potentialIssuers.length === 1) {
    return potentialIssuers[0];
  }
  if (potentialIssuers.length > 1) {
    debugLog("findIssuerCertificateInChain: certificate is not self-signed but has several issuers");
    return potentialIssuers[0];
  }
  return null;
}
var CertificateManagerState = /* @__PURE__ */ ((CertificateManagerState2) => {
  CertificateManagerState2[CertificateManagerState2["Uninitialized"] = 0] = "Uninitialized";
  CertificateManagerState2[CertificateManagerState2["Initializing"] = 1] = "Initializing";
  CertificateManagerState2[CertificateManagerState2["Initialized"] = 2] = "Initialized";
  CertificateManagerState2[CertificateManagerState2["Disposing"] = 3] = "Disposing";
  CertificateManagerState2[CertificateManagerState2["Disposed"] = 4] = "Disposed";
  return CertificateManagerState2;
})(CertificateManagerState || {});
var CertificateManager = class _CertificateManager extends import_node_events.EventEmitter {
  // ── Global instance registry ─────────────────────────────────
  // Tracks all initialized CertificateManager instances so their
  // file watchers can be closed automatically on process exit,
  // even if the consumer forgets to call dispose().
  static #activeInstances = /* @__PURE__ */ new Set();
  static #cleanupInstalled = false;
  static #installProcessCleanup() {
    if (_CertificateManager.#cleanupInstalled) return;
    _CertificateManager.#cleanupInstalled = true;
    const closeDanglingWatchers = () => {
      for (const cm of _CertificateManager.#activeInstances) {
        for (const w of cm.#watchers) {
          try {
            w.close();
          } catch {
          }
        }
        cm.#watchers.splice(0);
        cm.state = 4 /* Disposed */;
      }
      _CertificateManager.#activeInstances.clear();
    };
    process.on("beforeExit", closeDanglingWatchers);
    for (const signal of ["SIGINT", "SIGTERM"]) {
      process.once(signal, () => {
        closeDanglingWatchers();
        process.exit();
      });
    }
  }
  /**
   * Dispose **all** active CertificateManager instances,
   * closing their file watchers and freeing resources.
   *
   * This is mainly useful in test tear-down to ensure the
   * Node.js process can exit cleanly.
   */
  static async disposeAll() {
    const instances = [..._CertificateManager.#activeInstances];
    await Promise.all(instances.map((cm) => _CertificateManager.prototype.dispose.call(cm)));
  }
  /**
   * Assert that all CertificateManager instances have been
   * properly disposed. Throws an Error listing the locations
   * of any leaked instances.
   *
   * Intended for use in test `afterAll()` / `afterEach()`
   * hooks to catch missing `dispose()` calls early.
   *
   * @example
   * ```ts
   * after(() => {
   *     CertificateManager.checkAllDisposed();
   * });
   * ```
   */
  static checkAllDisposed() {
    if (_CertificateManager.#activeInstances.size === 0) return;
    const locations = [..._CertificateManager.#activeInstances].map((cm) => cm.rootDir);
    throw new Error(
      `${_CertificateManager.#activeInstances.size} CertificateManager instance(s) not disposed:
  - ${locations.join("\n  - ")}`
    );
  }
  // ─────────────────────────────────────────────────────────────
  /**
   * When `true` (the default), any certificate that is not
   * already in the trusted or rejected store is automatically
   * written to the rejected folder the first time it is seen.
   */
  untrustUnknownCertificate = true;
  /** Current lifecycle state of this instance. */
  state = 0 /* Uninitialized */;
  /** @deprecated Use {@link folderPollingInterval} instead (typo fix). */
  folderPoolingInterval = 5e3;
  /** Interval in milliseconds for file-system polling (when enabled). */
  get folderPollingInterval() {
    return this.folderPoolingInterval;
  }
  set folderPollingInterval(value) {
    this.folderPoolingInterval = value;
  }
  /** RSA key size used when generating the private key. */
  keySize;
  #location;
  #watchers = [];
  #pendingUnrefs = /* @__PURE__ */ new Set();
  #readCertificatesCalled = false;
  #filenameToHash = /* @__PURE__ */ new Map();
  #initializingPromise;
  #thumbs = {
    rejected: /* @__PURE__ */ new Map(),
    trusted: /* @__PURE__ */ new Map(),
    issuers: {
      certs: /* @__PURE__ */ new Map()
    },
    crl: /* @__PURE__ */ new Map(),
    issuersCrl: /* @__PURE__ */ new Map()
  };
  /**
   * Create a new CertificateManager.
   *
   * The constructor creates the root directory if it does not
   * exist but does **not** initialise the PKI store — call
   * {@link initialize} before using any other method.
   *
   * @param options - configuration options
   */
  constructor(options) {
    super();
    options.keySize = options.keySize || 2048;
    if (!options.location) {
      throw new Error("CertificateManager: missing 'location' option");
    }
    this.#location = makePath(options.location, "");
    this.keySize = options.keySize;
    mkdirRecursiveSync(options.location);
    if (!import_node_fs9.default.existsSync(this.#location)) {
      throw new Error(`CertificateManager cannot access location ${this.#location}`);
    }
  }
  /** Path to the OpenSSL configuration file. */
  get configFile() {
    return import_node_path6.default.join(this.rootDir, "own/openssl.cnf");
  }
  /** Root directory of the PKI store. */
  get rootDir() {
    return this.#location;
  }
  /** Path to the private key file (`own/private/private_key.pem`). */
  get privateKey() {
    return import_node_path6.default.join(this.rootDir, "own/private/private_key.pem");
  }
  /** Path to the OpenSSL random seed file. */
  get randomFile() {
    return import_node_path6.default.join(this.rootDir, "./random.rnd");
  }
  /**
   * Move a certificate to the rejected store.
   * If the certificate was previously trusted, it will be removed from the trusted folder.
   * @param certificate - the DER-encoded certificate
   */
  async rejectCertificate(certificate) {
    await this.#moveCertificate(certificate, "rejected");
  }
  /**
   * Move a certificate to the trusted store.
   * If the certificate was previously rejected, it will be removed from the rejected folder.
   * @param certificate - the DER-encoded certificate
   */
  async trustCertificate(certificate) {
    await this.#moveCertificate(certificate, "trusted");
  }
  /**
   * Check whether the trusted certificate store is empty.
   *
   * This inspects the in-memory index, which is kept in
   * sync with the `trusted/certs/` folder by file-system
   * watchers after {@link initialize} has been called.
   */
  isTrustListEmpty() {
    return this.#thumbs.trusted.size === 0;
  }
  /**
   * Return the number of certificates currently in the
   * trusted store.
   */
  getTrustedCertificateCount() {
    return this.#thumbs.trusted.size;
  }
  /** Path to the rejected certificates folder. */
  get rejectedFolder() {
    return import_node_path6.default.join(this.rootDir, "rejected");
  }
  /** Path to the trusted certificates folder. */
  get trustedFolder() {
    return import_node_path6.default.join(this.rootDir, "trusted/certs");
  }
  /** Path to the trusted CRL folder. */
  get crlFolder() {
    return import_node_path6.default.join(this.rootDir, "trusted/crl");
  }
  /** Path to the issuer (CA) certificates folder. */
  get issuersCertFolder() {
    return import_node_path6.default.join(this.rootDir, "issuers/certs");
  }
  /** Path to the issuer CRL folder. */
  get issuersCrlFolder() {
    return import_node_path6.default.join(this.rootDir, "issuers/crl");
  }
  /** Path to the own certificate folder. */
  get ownCertFolder() {
    return import_node_path6.default.join(this.rootDir, "own/certs");
  }
  get ownPrivateFolder() {
    return import_node_path6.default.join(this.rootDir, "own/private");
  }
  /**
   * Check if a certificate is in the trusted store.
   * If the certificate is unknown and `untrustUnknownCertificate` is set,
   * it will be written to the rejected folder.
   * @param certificate - the DER-encoded certificate
   * @returns `"Good"` if trusted, `"BadCertificateUntrusted"` if rejected/unknown,
   *   or `"BadCertificateInvalid"` if the certificate cannot be parsed.
   */
  async isCertificateTrusted(certificate) {
    const fingerprint = makeFingerprint(certificate);
    if (this.#thumbs.trusted.has(fingerprint)) {
      return "Good";
    }
    if (!this.#thumbs.rejected.has(fingerprint)) {
      if (!this.untrustUnknownCertificate) {
        return "Good";
      }
      try {
        (0, import_node_opcua_crypto5.exploreCertificateInfo)(certificate);
      } catch (_err) {
        return "BadCertificateInvalid";
      }
      const filename = import_node_path6.default.join(this.rejectedFolder, `${buildIdealCertificateName(certificate)}.pem`);
      debugLog("certificate has never been seen before and is now rejected (untrusted) ", filename);
      await fsWriteFile(filename, (0, import_node_opcua_crypto5.toPem)(certificate, "CERTIFICATE"));
      this.#thumbs.rejected.set(fingerprint, { certificate, filename });
    }
    return "BadCertificateUntrusted";
  }
  async #innerVerifyCertificateAsync(certificate, _isIssuer, level, options) {
    if (level >= 5) {
      return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
    }
    const chain = (0, import_node_opcua_crypto5.split_der)(certificate);
    debugLog("NB CERTIFICATE IN CHAIN = ", chain.length);
    const info = exploreCertificateCached(chain[0]);
    let hasValidIssuer = false;
    let hasTrustedIssuer = false;
    const hasIssuerKey = info.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
    debugLog("Certificate as an Issuer Key", hasIssuerKey);
    if (hasIssuerKey) {
      const isSelfSigned = isSelfSigned2(info);
      debugLog("Is the Certificate self-signed  ?", isSelfSigned);
      if (!isSelfSigned) {
        debugLog(
          "Is issuer found in the list of know issuers ?",
          "\n subjectKeyIdentifier   = ",
          info.tbsCertificate.extensions?.subjectKeyIdentifier,
          "\n authorityKeyIdentifier = ",
          info.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier
        );
        let issuerCertificate = await this.findIssuerCertificate(chain[0]);
        if (!issuerCertificate) {
          issuerCertificate = findIssuerCertificateInChain(chain[0], chain);
          if (!issuerCertificate) {
            debugLog(
              " the issuer has not been found in the chain itself nor in the issuer.cert list => the chain is incomplete!"
            );
            return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
          }
          debugLog(" the issuer certificate has been found in the chain itself ! the chain is complete !");
        } else {
          debugLog(" the issuer certificate has been found in the issuer.cert folder !");
        }
        const issuerStatus = await this.#innerVerifyCertificateAsync(issuerCertificate, true, level + 1, options);
        if (issuerStatus === "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */) {
          return "BadCertificateIssuerRevocationUnknown" /* BadCertificateIssuerRevocationUnknown */;
        }
        if (issuerStatus === "BadCertificateIssuerRevocationUnknown" /* BadCertificateIssuerRevocationUnknown */) {
          return "BadCertificateIssuerRevocationUnknown" /* BadCertificateIssuerRevocationUnknown */;
        }
        if (issuerStatus === "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */) {
          if (!options || !options.acceptOutDatedIssuerCertificate) {
            return "BadCertificateIssuerTimeInvalid" /* BadCertificateIssuerTimeInvalid */;
          }
        }
        if (issuerStatus === "BadCertificateUntrusted" /* BadCertificateUntrusted */) {
          debugLog("warning issuerStatus = ", issuerStatus.toString(), "the issuer certificate is not trusted");
        }
        if (issuerStatus !== "Good" /* Good */ && issuerStatus !== "BadCertificateUntrusted" /* BadCertificateUntrusted */) {
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
        const isCertificateSignatureOK = (0, import_node_opcua_crypto5.verifyCertificateSignature)(certificate, issuerCertificate);
        if (!isCertificateSignatureOK) {
          debugLog(" the certificate was not signed by the issuer as it claim to be ! Danger");
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
        hasValidIssuer = true;
        let revokedStatus = await this.isCertificateRevoked(certificate);
        if (revokedStatus === "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */) {
          if (options?.ignoreMissingRevocationList) {
            revokedStatus = "Good" /* Good */;
          }
        }
        if (revokedStatus !== "Good" /* Good */) {
          debugLog("revokedStatus", revokedStatus);
          return revokedStatus;
        }
        const issuerTrustedStatus = await this.#checkRejectedOrTrusted(issuerCertificate);
        debugLog("issuerTrustedStatus", issuerTrustedStatus);
        if (issuerTrustedStatus === "unknown") {
          hasTrustedIssuer = false;
        } else if (issuerTrustedStatus === "trusted") {
          hasTrustedIssuer = true;
        } else if (issuerTrustedStatus === "rejected") {
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
      } else {
        const isCertificateSignatureOK = (0, import_node_opcua_crypto5.verifyCertificateSignature)(certificate, certificate);
        if (!isCertificateSignatureOK) {
          debugLog("Self-signed Certificate signature is not valid");
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
        const revokedStatus = await this.isCertificateRevoked(certificate);
        debugLog("revokedStatus of self signed certificate:", revokedStatus);
      }
    }
    const status = await this.#checkRejectedOrTrusted(certificate);
    if (status === "rejected") {
      if (!(options.acceptCertificateWithValidIssuerChain && hasValidIssuer && hasTrustedIssuer)) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
    }
    const _c2 = chain[1] ? (0, import_node_opcua_crypto5.exploreCertificateInfo)(chain[1]) : "non";
    debugLog("chain[1] info=", _c2);
    const certificateInfo = (0, import_node_opcua_crypto5.exploreCertificateInfo)(certificate);
    const now = /* @__PURE__ */ new Date();
    let isTimeInvalid = false;
    if (certificateInfo.notBefore.getTime() > now.getTime()) {
      debugLog(
        `${import_chalk6.default.red("certificate is invalid : certificate is not active yet !")} not before date =${certificateInfo.notBefore}`
      );
      if (!options.acceptPendingCertificate) {
        isTimeInvalid = true;
      }
    }
    if (certificateInfo.notAfter.getTime() <= now.getTime()) {
      debugLog(
        `${import_chalk6.default.red("certificate is invalid : certificate has expired !")} not after date =${certificateInfo.notAfter}`
      );
      if (!options.acceptOutdatedCertificate) {
        isTimeInvalid = true;
      }
    }
    if (status === "trusted") {
      return isTimeInvalid ? "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */ : "Good" /* Good */;
    }
    if (hasIssuerKey) {
      if (!hasTrustedIssuer) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
      if (!hasValidIssuer) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
      if (!options.acceptCertificateWithValidIssuerChain) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
      return isTimeInvalid ? "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */ : "Good" /* Good */;
    } else {
      return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
    }
  }
  /**
   * Internal verification hook called by {@link verifyCertificate}.
   *
   * Subclasses can override this to inject additional validation
   * logic (e.g. application-level policy checks) while still
   * delegating to the default chain/CRL/trust verification.
   *
   * @param certificate - the DER-encoded certificate to verify
   * @param options - verification options forwarded from the
   *   public API
   * @returns the verification status code
   */
  async verifyCertificateAsync(certificate, options) {
    const status1 = await this.#innerVerifyCertificateAsync(certificate, false, 0, options);
    return status1;
  }
  /**
   * Verify a certificate against the PKI trust store.
   *
   * This performs a full validation including trust status,
   * issuer chain, CRL revocation checks, and time validity.
   *
   * @param certificate - the DER-encoded certificate to verify
   * @param options - optional flags to relax validation rules
   * @returns the verification status code
   */
  async verifyCertificate(certificate, options) {
    if (!certificate) {
      return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
    }
    try {
      const status = await this.verifyCertificateAsync(certificate, options || {});
      return status;
    } catch (error) {
      warningLog(`verifyCertificate error: ${error.message}`);
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
  }
  /**
   * Initialize the PKI directory structure, generate the
   * private key (if missing), and start file-system watchers.
   *
   * This method is idempotent — subsequent calls are no-ops.
   * It must be called before any certificate operations.
   */
  async initialize() {
    if (this.state !== 0 /* Uninitialized */) {
      return;
    }
    this.state = 1 /* Initializing */;
    this.#initializingPromise = this.#initialize();
    await this.#initializingPromise;
    this.#initializingPromise = void 0;
    this.state = 2 /* Initialized */;
    _CertificateManager.#activeInstances.add(this);
    _CertificateManager.#installProcessCleanup();
  }
  async #initialize() {
    this.state = 1 /* Initializing */;
    const pkiDir = this.#location;
    mkdirRecursiveSync(pkiDir);
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "own"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "own/certs"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "own/private"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "rejected"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "trusted"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "trusted/certs"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "trusted/crl"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "issuers"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "issuers/certs"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "issuers/crl"));
    if (!import_node_fs9.default.existsSync(this.configFile) || !import_node_fs9.default.existsSync(this.privateKey)) {
      return await this.withLock2(async () => {
        if (this.state === 3 /* Disposing */ || this.state === 4 /* Disposed */) {
          return;
        }
        if (!import_node_fs9.default.existsSync(this.configFile)) {
          import_node_fs9.default.writeFileSync(this.configFile, configurationFileSimpleTemplate);
        }
        if (!import_node_fs9.default.existsSync(this.privateKey)) {
          debugLog("generating private key ...");
          await (0, import_node_opcua_crypto5.generatePrivateKeyFile)(this.privateKey, this.keySize);
          await this.#readCertificates();
        } else {
          await this.#readCertificates();
        }
      });
    } else {
      await this.#readCertificates();
    }
  }
  /**
   * Dispose of the CertificateManager, releasing file watchers
   * and other resources. The instance should not be used after
   * calling this method.
   */
  async dispose() {
    if (this.state === 3 /* Disposing */) {
      throw new Error("Already disposing");
    }
    if (this.state === 0 /* Uninitialized */) {
      this.state = 4 /* Disposed */;
      return;
    }
    if (this.state === 1 /* Initializing */) {
      if (this.#initializingPromise) {
        await this.#initializingPromise;
      }
    }
    try {
      this.state = 3 /* Disposing */;
      await (0, import_global_mutex.drainPendingLocks)();
      for (const unreff of this.#pendingUnrefs) {
        unreff();
      }
      this.#pendingUnrefs.clear();
      await Promise.all(this.#watchers.map((w) => w.close()));
      this.#watchers.forEach((w) => {
        w.removeAllListeners();
      });
      this.#watchers.splice(0);
    } finally {
      this.state = 4 /* Disposed */;
      _CertificateManager.#activeInstances.delete(this);
    }
  }
  /**
   * Force a full re-scan of all PKI folders, rebuilding
   * the in-memory `_thumbs` index from scratch.
   *
   * Call this after external processes have modified the
   * PKI folders (e.g. via `writeTrustList` or CLI tools)
   * to ensure the CertificateManager sees the latest
   * state without waiting for file-system events.
   */
  async reloadCertificates() {
    await Promise.all(this.#watchers.map((w) => w.close()));
    for (const w of this.#watchers) {
      w.removeAllListeners();
    }
    this.#watchers.splice(0);
    this.#thumbs.rejected.clear();
    this.#thumbs.trusted.clear();
    this.#thumbs.issuers.certs.clear();
    this.#thumbs.crl.clear();
    this.#thumbs.issuersCrl.clear();
    this.#filenameToHash.clear();
    this.#readCertificatesCalled = false;
    await this.#readCertificates();
  }
  async withLock2(action) {
    const lockFileName = import_node_path6.default.join(this.rootDir, "mutex");
    return (0, import_global_mutex.withLock)({ fileToLock: lockFileName }, async () => {
      return await action();
    });
  }
  /**
   * Create a self-signed certificate for this PKI's private key.
   *
   * The certificate is written to `params.outputFile` or
   * `own/certs/self_signed_certificate.pem` by default.
   *
   * @param params - certificate parameters (subject, SANs,
   *   validity, etc.)
   */
  async createSelfSignedCertificate(params) {
    if (typeof params.applicationUri !== "string") {
      throw new Error("createSelfSignedCertificate: expecting applicationUri to be a string");
    }
    if (!import_node_fs9.default.existsSync(this.privateKey)) {
      throw new Error(`Cannot find private key ${this.privateKey}`);
    }
    let certificateFilename = import_node_path6.default.join(this.rootDir, "own/certs/self_signed_certificate.pem");
    certificateFilename = params.outputFile || certificateFilename;
    const _params = params;
    _params.rootDir = this.rootDir;
    _params.configFile = this.configFile;
    _params.privateKey = this.privateKey;
    _params.subject = params.subject || "CN=FIXME";
    await this.withLock2(async () => {
      await createSelfSignedCertificate(certificateFilename, _params);
    });
  }
  /**
   * Create a Certificate Signing Request (CSR) using this
   * PKI's private key and configuration.
   *
   * The CSR file is written to `own/certs/` with a timestamped
   * filename.
   *
   * @param params - CSR parameters (subject, SANs)
   * @returns the filesystem path to the generated CSR file
   */
  async createCertificateRequest(params) {
    if (!params) {
      throw new Error("params is required");
    }
    const _params = params;
    if (Object.prototype.hasOwnProperty.call(_params, "rootDir")) {
      throw new Error("rootDir should not be specified ");
    }
    _params.rootDir = import_node_path6.default.resolve(this.rootDir);
    _params.configFile = import_node_path6.default.resolve(this.configFile);
    _params.privateKey = import_node_path6.default.resolve(this.privateKey);
    return await this.withLock2(async () => {
      const now = /* @__PURE__ */ new Date();
      const today = `${now.toISOString().slice(0, 10)}_${now.getTime()}`;
      const certificateSigningRequestFilename = import_node_path6.default.join(this.rootDir, "own/certs", `certificate_${today}.csr`);
      await createCertificateSigningRequestAsync(certificateSigningRequestFilename, _params);
      return certificateSigningRequestFilename;
    });
  }
  /**
   * Add a CA (issuer) certificate to the issuers store.
   * If the certificate is already present, this is a no-op.
   * @param certificate - the DER-encoded CA certificate
   * @param validate - if `true`, verify the certificate before adding
   * @param addInTrustList - if `true`, also add to the trusted store
   * @returns `VerificationStatus.Good` on success
   */
  async addIssuer(certificate, validate = false, addInTrustList = false) {
    if (validate) {
      const status = await this.verifyCertificate(certificate);
      if (status !== "Good" /* Good */ && status !== "BadCertificateUntrusted" /* BadCertificateUntrusted */) {
        return status;
      }
    }
    const pemCertificate = (0, import_node_opcua_crypto5.toPem)(certificate, "CERTIFICATE");
    const fingerprint = makeFingerprint(certificate);
    if (this.#thumbs.issuers.certs.has(fingerprint)) {
      return "Good" /* Good */;
    }
    const filename = import_node_path6.default.join(this.issuersCertFolder, `issuer_${buildIdealCertificateName(certificate)}.pem`);
    await import_node_fs9.default.promises.writeFile(filename, pemCertificate, "ascii");
    this.#thumbs.issuers.certs.set(fingerprint, { certificate, filename });
    if (addInTrustList) {
      await this.trustCertificate(certificate);
    }
    return "Good" /* Good */;
  }
  /**
   * Add a CRL to the certificate manager.
   * @param crl - the CRL to add
   * @param target - "issuers" (default) writes to issuers/crl, "trusted" writes to trusted/crl
   */
  async addRevocationList(crl, target = "issuers") {
    return await this.withLock2(async () => {
      try {
        const index = target === "trusted" ? this.#thumbs.crl : this.#thumbs.issuersCrl;
        const folder = target === "trusted" ? this.crlFolder : this.issuersCrlFolder;
        const crlInfo = (0, import_node_opcua_crypto5.exploreCertificateRevocationList)(crl);
        const key = crlInfo.tbsCertList.issuerFingerprint;
        if (!index.has(key)) {
          index.set(key, { crls: [], serialNumbers: {} });
        }
        const pemCertificate = (0, import_node_opcua_crypto5.toPem)(crl, "X509 CRL");
        const filename = import_node_path6.default.join(folder, `crl_${buildIdealCertificateName(crl)}.pem`);
        await import_node_fs9.default.promises.writeFile(filename, pemCertificate, "ascii");
        await this.#onCrlFileAdded(index, filename);
        await this.#waitAndCheckCRLProcessingStatus();
        return "Good" /* Good */;
      } catch (err) {
        debugLog(err);
        return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
      }
    });
  }
  /**
   * Remove all CRL files from the specified folder(s) and clear the
   * corresponding in-memory index.
   * @param target - "issuers" clears issuers/crl, "trusted" clears
   *   trusted/crl, "all" clears both.
   */
  async clearRevocationLists(target) {
    const clearFolder = async (folder, index) => {
      try {
        const files = await import_node_fs9.default.promises.readdir(folder);
        for (const file of files) {
          const ext = import_node_path6.default.extname(file).toLowerCase();
          if (ext === ".crl" || ext === ".pem" || ext === ".der") {
            await import_node_fs9.default.promises.unlink(import_node_path6.default.join(folder, file));
          }
        }
      } catch (err) {
        if (err.code !== "ENOENT") {
          throw err;
        }
      }
      index.clear();
    };
    if (target === "issuers" || target === "all") {
      await clearFolder(this.issuersCrlFolder, this.#thumbs.issuersCrl);
    }
    if (target === "trusted" || target === "all") {
      await clearFolder(this.crlFolder, this.#thumbs.crl);
    }
  }
  /**
   * Check whether an issuer certificate with the given thumbprint
   * is already registered.
   * @param thumbprint - hex-encoded SHA-1 thumbprint (lowercase)
   */
  async hasIssuer(thumbprint) {
    await this.#readCertificates();
    const normalized = thumbprint.toLowerCase();
    return this.#thumbs.issuers.certs.has(normalized);
  }
  /**
   * Remove a trusted certificate identified by its SHA-1 thumbprint.
   * Deletes the file on disk and removes the entry from the
   * in-memory index.
   * @param thumbprint - hex-encoded SHA-1 thumbprint (lowercase)
   * @returns the removed certificate buffer, or `null` if not found
   */
  async removeTrustedCertificate(thumbprint) {
    await this.#readCertificates();
    const normalized = thumbprint.toLowerCase();
    const entry = this.#thumbs.trusted.get(normalized);
    if (!entry) {
      return null;
    }
    try {
      await import_node_fs9.default.promises.unlink(entry.filename);
    } catch (err) {
      if (err.code !== "ENOENT") {
        throw err;
      }
    }
    this.#thumbs.trusted.delete(normalized);
    return entry.certificate;
  }
  /**
   * Remove an issuer certificate identified by its SHA-1 thumbprint.
   * Deletes the file on disk and removes the entry from the
   * in-memory index.
   * @param thumbprint - hex-encoded SHA-1 thumbprint (lowercase)
   * @returns the removed certificate buffer, or `null` if not found
   */
  async removeIssuer(thumbprint) {
    await this.#readCertificates();
    const normalized = thumbprint.toLowerCase();
    const entry = this.#thumbs.issuers.certs.get(normalized);
    if (!entry) {
      return null;
    }
    try {
      await import_node_fs9.default.promises.unlink(entry.filename);
    } catch (err) {
      if (err.code !== "ENOENT") {
        throw err;
      }
    }
    this.#thumbs.issuers.certs.delete(normalized);
    return entry.certificate;
  }
  /**
   * Remove all CRL files that were issued by the given CA certificate
   * from the specified folder (or both).
   * @param issuerCertificate - the CA certificate whose CRLs to remove
   * @param target - "issuers", "trusted", or "all" (default "all")
   */
  async removeRevocationListsForIssuer(issuerCertificate, target = "all") {
    const issuerInfo = exploreCertificateCached(issuerCertificate);
    const issuerFingerprint = issuerInfo.tbsCertificate.subjectFingerPrint;
    const processIndex = async (index) => {
      const crlData = index.get(issuerFingerprint);
      if (!crlData) return;
      for (const crlEntry of crlData.crls) {
        try {
          await import_node_fs9.default.promises.unlink(crlEntry.filename);
        } catch (err) {
          if (err.code !== "ENOENT") {
            throw err;
          }
        }
      }
      index.delete(issuerFingerprint);
    };
    if (target === "issuers" || target === "all") {
      await processIndex(this.#thumbs.issuersCrl);
    }
    if (target === "trusted" || target === "all") {
      await processIndex(this.#thumbs.crl);
    }
  }
  /**
   * Validate a certificate (optionally with its chain) and add
   * the leaf certificate to the trusted store.
   *
   * The certificate buffer may contain a single certificate or a
   * full chain (leaf + issuer certificates concatenated in DER).
   * Only the leaf certificate is added to the trusted store.
   *
   * When the chain contains issuer certificates, this method
   * verifies that each issuer is already registered via
   * {@link addIssuer} before trusting the leaf.
   *
   * If one of the certificates in the chain is not registered in the issuers store,
   * the leaf certificate will be rejected.
   *
   * @param certificateChain - DER-encoded certificate or chain
   * @returns `VerificationStatus.Good` on success, or an error
   *  status indicating why the certificate was rejected.
   */
  async addTrustedCertificateFromChain(certificateChain) {
    const certificates = (0, import_node_opcua_crypto5.split_der)(certificateChain);
    const leafCertificate = certificates[0];
    try {
      exploreCertificateCached(leafCertificate);
    } catch (_err) {
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
    const result = await (0, import_node_opcua_crypto5.verifyCertificateChain)([leafCertificate]);
    if (result.status !== "Good") {
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
    if (certificates.length > 1) {
      await this.#scanCertFolder(this.issuersCertFolder, this.#thumbs.issuers.certs);
      for (const issuerCert of certificates.slice(1)) {
        const thumbprint = makeFingerprint(issuerCert);
        if (!await this.hasIssuer(thumbprint)) {
          return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
        }
      }
    }
    await this.trustCertificate(leafCertificate);
    return "Good" /* Good */;
  }
  /**
   * Check whether an issuer certificate is still needed by any
   * certificate in the trusted store.
   *
   * This is used before removing an issuer to ensure that
   * doing so would not break the chain of any trusted
   * certificate.
   *
   * @param issuerCertificate - the CA certificate to check
   * @returns `true` if at least one trusted certificate was
   *   signed by this issuer.
   */
  async isIssuerInUseByTrustedCertificate(issuerCertificate) {
    await this.#readCertificates();
    for (const entry of this.#thumbs.trusted.values()) {
      if (!entry.certificate) continue;
      try {
        if ((0, import_node_opcua_crypto5.verifyCertificateSignature)(entry.certificate, issuerCertificate)) {
          return true;
        }
      } catch (_err) {
      }
    }
    return false;
  }
  /**
   *  find the issuer certificate among the trusted  issuer certificates.
   *
   *  The findIssuerCertificate method is an asynchronous method that attempts to find
   *  the issuer certificate for a given certificate from the list of issuer certificate declared in the PKI
   *
   *  - If the certificate is self-signed, it returns the certificate itself.
   *
   *  - If the certificate has no extension 3, it is assumed to be generated by an old system, and a null value is returned.
   *
   *  - the method checks both issuer and trusted certificates and returns the appropriate issuercertificate,
   *    if found. If multiple matching certificates are found, a warning is logged to the console.
   *
   */
  async findIssuerCertificate(certificate) {
    const certInfo = exploreCertificateCached(certificate);
    if (isSelfSigned2(certInfo)) {
      return certificate;
    }
    const wantedIssuerKey = certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
    if (!wantedIssuerKey) {
      debugLog("Certificate has no extension 3");
      return null;
    }
    const issuerCertificates = [...this.#thumbs.issuers.certs.values()];
    const selectedIssuerCertificates = findMatchingIssuerKey(issuerCertificates, wantedIssuerKey);
    if (selectedIssuerCertificates.length > 0) {
      if (selectedIssuerCertificates.length > 1) {
        warningLog("Warning more than one issuer certificate exists with subjectKeyIdentifier ", wantedIssuerKey);
      }
      return selectedIssuerCertificates[0].certificate || null;
    }
    const trustedCertificates = [...this.#thumbs.trusted.values()];
    const selectedTrustedCertificates = findMatchingIssuerKey(trustedCertificates, wantedIssuerKey);
    if (selectedTrustedCertificates.length > 1) {
      warningLog(
        "Warning more than one certificate exists with subjectKeyIdentifier in trusted certificate list ",
        wantedIssuerKey,
        selectedTrustedCertificates.length
      );
    }
    return selectedTrustedCertificates.length > 0 ? selectedTrustedCertificates[0].certificate : null;
  }
  /**
   *
   * check if the certificate explicitly appear in the trust list, the reject list or none.
   * In case of being in the reject and trusted list at the same time is consider: rejected.
   * @internal
   * @private
   */
  async #checkRejectedOrTrusted(certificate) {
    const fingerprint = makeFingerprint(certificate);
    debugLog("#checkRejectedOrTrusted fingerprint ", short(fingerprint));
    await this.#readCertificates();
    if (this.#thumbs.rejected.has(fingerprint)) {
      return "rejected";
    }
    if (this.#thumbs.trusted.has(fingerprint)) {
      return "trusted";
    }
    return "unknown";
  }
  async #moveCertificate(certificate, newStatus) {
    await this.withLock2(async () => {
      const fingerprint = makeFingerprint(certificate);
      let status = await this.#checkRejectedOrTrusted(certificate);
      if (status === "unknown") {
        const pem = (0, import_node_opcua_crypto5.toPem)(certificate, "CERTIFICATE");
        const filename = import_node_path6.default.join(this.rejectedFolder, `${buildIdealCertificateName(certificate)}.pem`);
        await import_node_fs9.default.promises.writeFile(filename, pem);
        this.#thumbs.rejected.set(fingerprint, { certificate, filename });
        status = "rejected";
      }
      debugLog("#moveCertificate", fingerprint.substring(0, 10), "from", status, "to", newStatus);
      if (status !== "rejected" && status !== "trusted") {
        throw new Error(`#moveCertificate: unexpected status '${status}' for certificate ${fingerprint.substring(0, 10)}`);
      }
      if (status !== newStatus) {
        const indexSrc = status === "rejected" ? this.#thumbs.rejected : this.#thumbs.trusted;
        const srcEntry = indexSrc.get(fingerprint);
        if (!srcEntry) {
          debugLog(" cannot find certificate ", fingerprint.substring(0, 10), " in", status);
          throw new Error(`#moveCertificate: certificate ${fingerprint.substring(0, 10)} not found in ${status} index`);
        }
        const destFolder = newStatus === "trusted" ? this.trustedFolder : this.rejectedFolder;
        const certificateDest = import_node_path6.default.join(destFolder, import_node_path6.default.basename(srcEntry.filename));
        debugLog("#moveCertificate", fingerprint.substring(0, 10), "old name", srcEntry.filename);
        debugLog("#moveCertificate", fingerprint.substring(0, 10), "new name", certificateDest);
        await import_node_fs9.default.promises.rename(srcEntry.filename, certificateDest);
        indexSrc.delete(fingerprint);
        const indexDest = newStatus === "trusted" ? this.#thumbs.trusted : this.#thumbs.rejected;
        indexDest.set(fingerprint, { certificate, filename: certificateDest });
      }
    });
  }
  #findAssociatedCRLs(issuerCertificate) {
    const issuerCertificateInfo = exploreCertificateCached(issuerCertificate);
    const key = issuerCertificateInfo.tbsCertificate.subjectFingerPrint;
    return this.#thumbs.issuersCrl.get(key) ?? this.#thumbs.crl.get(key) ?? null;
  }
  /**
   * Check whether a certificate has been revoked by its issuer's CRL.
   *
   * - Self-signed certificates are never considered revoked.
   * - If no `issuerCertificate` is provided, the method attempts
   *   to find it via {@link findIssuerCertificate}.
   *
   * @param certificate - the DER-encoded certificate to check
   * @param issuerCertificate - optional issuer certificate; looked
   *   up automatically when omitted
   * @returns `Good` if not revoked, `BadCertificateRevoked` if the
   *   serial number appears in a CRL,
   *   `BadCertificateRevocationUnknown` if no CRL is available,
   *   or `BadCertificateChainIncomplete` if the issuer cannot be
   *   found.
   */
  async isCertificateRevoked(certificate, issuerCertificate) {
    if (isSelfSigned3(certificate)) {
      return "Good" /* Good */;
    }
    if (!issuerCertificate) {
      issuerCertificate = await this.findIssuerCertificate(certificate);
    }
    if (!issuerCertificate) {
      return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
    }
    const crls = this.#findAssociatedCRLs(issuerCertificate);
    if (!crls) {
      return "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */;
    }
    const certInfo = exploreCertificateCached(certificate);
    const serialNumber = certInfo.tbsCertificate.serialNumber || certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.serial || "";
    const key = certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.authorityCertIssuerFingerPrint || "<unknown>";
    const crl2 = this.#thumbs.crl.get(key) ?? null;
    if (crls.serialNumbers[serialNumber] || crl2?.serialNumbers[serialNumber]) {
      return "BadCertificateRevoked" /* BadCertificateRevoked */;
    }
    return "Good" /* Good */;
  }
  #pendingCrlToProcess = 0;
  #onCrlProcessWaiters = [];
  #queue = [];
  #onCrlFileAdded(index, filename) {
    this.#queue.push({ index, filename });
    this.#pendingCrlToProcess += 1;
    if (this.#pendingCrlToProcess === 1) {
      this.#processNextCrl();
    }
  }
  async #processNextCrl() {
    try {
      const nextCRL = this.#queue.shift();
      if (!nextCRL) return;
      const { index, filename } = nextCRL;
      const crl = await (0, import_node_opcua_crypto5.readCertificateRevocationList)(filename);
      const crlInfo = (0, import_node_opcua_crypto5.exploreCertificateRevocationList)(crl);
      debugLog(import_chalk6.default.cyan("add CRL in folder "), filename);
      const fingerprint = crlInfo.tbsCertList.issuerFingerprint;
      if (!index.has(fingerprint)) {
        index.set(fingerprint, { crls: [], serialNumbers: {} });
      }
      const data = index.get(fingerprint) || { crls: [], serialNumbers: {} };
      data.crls.push({ crlInfo, filename });
      for (const revokedCertificate of crlInfo.tbsCertList.revokedCertificates) {
        const serialNumber = revokedCertificate.userCertificate;
        if (!data.serialNumbers[serialNumber]) {
          data.serialNumbers[serialNumber] = revokedCertificate.revocationDate;
        }
      }
      debugLog(import_chalk6.default.cyan("CRL"), fingerprint, "serial numbers = ", Object.keys(data.serialNumbers));
    } catch (err) {
      debugLog("CRL filename error =");
      debugLog(err);
    }
    this.#pendingCrlToProcess -= 1;
    if (this.#pendingCrlToProcess === 0) {
      for (const waiter of this.#onCrlProcessWaiters) {
        waiter();
      }
      this.#onCrlProcessWaiters.length = 0;
    } else {
      this.#processNextCrl();
    }
  }
  async #readCertificates() {
    if (this.#readCertificatesCalled) {
      return;
    }
    this.#readCertificatesCalled = true;
    const usePolling = process.env.OPCUA_PKI_USE_POLLING === "true";
    const envInterval = process.env.OPCUA_PKI_POLLING_INTERVAL ? parseInt(process.env.OPCUA_PKI_POLLING_INTERVAL, 10) : void 0;
    const pollingInterval = Math.min(10 * 60 * 1e3, Math.max(100, envInterval ?? this.folderPollingInterval));
    const chokidarOptions = {
      usePolling,
      ...usePolling ? { interval: pollingInterval } : {},
      persistent: false
    };
    const allCapturedHandles = [];
    const origWatch = import_node_fs9.default.watch;
    let watcherReadyCount = 0;
    const totalWatchers = 5;
    import_node_fs9.default.watch = ((...args) => {
      const handle = origWatch.apply(import_node_fs9.default, args);
      handle.setMaxListeners(handle.getMaxListeners() + 1);
      handle.on("error", () => {
      });
      allCapturedHandles.push(handle);
      return handle;
    });
    const createUnreffedWatcher = (folder) => {
      const startIdx = allCapturedHandles.length;
      const w = import_chokidar.default.watch(folder, chokidarOptions);
      const unreffAll = () => {
        for (let i = startIdx; i < allCapturedHandles.length; i++) {
          allCapturedHandles[i].unref();
        }
        watcherReadyCount++;
        if (watcherReadyCount >= totalWatchers) {
          import_node_fs9.default.watch = origWatch;
        }
      };
      return { w, capturedHandles: allCapturedHandles.slice(startIdx), unreffAll };
    };
    await Promise.all([
      this.#scanCertFolder(this.trustedFolder, this.#thumbs.trusted),
      this.#scanCertFolder(this.issuersCertFolder, this.#thumbs.issuers.certs),
      this.#scanCertFolder(this.rejectedFolder, this.#thumbs.rejected),
      this.#scanCrlFolder(this.crlFolder, this.#thumbs.crl),
      this.#scanCrlFolder(this.issuersCrlFolder, this.#thumbs.issuersCrl)
    ]);
    this.#startWatcher(this.trustedFolder, this.#thumbs.trusted, createUnreffedWatcher, "trusted");
    this.#startWatcher(this.issuersCertFolder, this.#thumbs.issuers.certs, createUnreffedWatcher, "issuersCerts");
    this.#startWatcher(this.rejectedFolder, this.#thumbs.rejected, createUnreffedWatcher, "rejected");
    this.#startCrlWatcher(this.crlFolder, this.#thumbs.crl, createUnreffedWatcher, "crl");
    this.#startCrlWatcher(this.issuersCrlFolder, this.#thumbs.issuersCrl, createUnreffedWatcher, "issuersCrl");
  }
  /**
   * Scan a certificate folder and populate the in-memory index.
   * Uses async readdir/stat to yield the event loop between
   * file reads, preventing main-loop stalls with large folders.
   */
  async #scanCertFolder(folder, index) {
    if (!import_node_fs9.default.existsSync(folder)) return;
    const files = await import_node_fs9.default.promises.readdir(folder);
    for (const file of files) {
      const filename = import_node_path6.default.join(folder, file);
      try {
        const stat = await import_node_fs9.default.promises.stat(filename);
        if (!stat.isFile()) continue;
        const certificate = await (0, import_node_opcua_crypto5.readCertificateAsync)(filename);
        const info = exploreCertificateCached(certificate);
        const fingerprint = makeFingerprint(certificate);
        index.set(fingerprint, { certificate, filename, info });
        this.#filenameToHash.set(filename, fingerprint);
      } catch (err) {
        debugLog(`scanCertFolder: skipping ${filename}`, err);
      }
    }
  }
  /**
   * Scan a CRL folder and populate the in-memory CRL index.
   */
  async #scanCrlFolder(folder, index) {
    if (!import_node_fs9.default.existsSync(folder)) return;
    const files = await import_node_fs9.default.promises.readdir(folder);
    for (const file of files) {
      const filename = import_node_path6.default.join(folder, file);
      try {
        const stat = await import_node_fs9.default.promises.stat(filename);
        if (!stat.isFile()) continue;
        this.#onCrlFileAdded(index, filename);
      } catch (err) {
        debugLog(`scanCrlFolder: skipping ${filename}`, err);
      }
    }
    await this.#waitAndCheckCRLProcessingStatus();
  }
  /**
   * Start a chokidar watcher for a CRL folder.
   * Non-blocking — does NOT await "ready".
   */
  #startCrlWatcher(folder, index, createUnreffedWatcher, store) {
    const { w, unreffAll } = createUnreffedWatcher(folder);
    w.on("error", (err) => {
      debugLog(`chokidar CRL watcher error on ${folder}:`, err);
    });
    let ready = false;
    w.on("unlink", (filename) => {
      for (const [key, data] of index.entries()) {
        data.crls = data.crls.filter((c) => c.filename !== filename);
        if (data.crls.length === 0) {
          index.delete(key);
        }
      }
      if (ready) {
        this.emit("crlRemoved", { store, filename });
      }
    });
    w.on("add", (filename) => {
      if (ready) {
        this.#onCrlFileAdded(index, filename);
        this.emit("crlAdded", { store, filename });
      }
    });
    w.on("change", (changedPath) => {
      debugLog("change in folder ", folder, changedPath);
    });
    this.#watchers.push(w);
    this.#pendingUnrefs.add(unreffAll);
    w.on("ready", () => {
      ready = true;
      this.#pendingUnrefs.delete(unreffAll);
      unreffAll();
    });
  }
  /**
   * Start a chokidar watcher for a certificate folder.
   * Non-blocking — does NOT await "ready".
   */
  #startWatcher(folder, index, createUnreffedWatcher, store) {
    const { w, unreffAll } = createUnreffedWatcher(folder);
    w.on("error", (err) => {
      debugLog(`chokidar cert watcher error on ${folder}:`, err);
    });
    let ready = false;
    w.on("unlink", (filename) => {
      debugLog(import_chalk6.default.cyan(`unlink in folder ${folder}`), filename);
      const h = this.#filenameToHash.get(filename);
      if (h && index.has(h)) {
        index.delete(h);
        this.emit("certificateRemoved", { store, fingerprint: h, filename });
      }
    });
    w.on("add", (filename) => {
      debugLog(import_chalk6.default.cyan(`add in folder ${folder}`), filename);
      try {
        const certificate = (0, import_node_opcua_crypto5.readCertificate)(filename);
        const info = exploreCertificateCached(certificate);
        const fingerprint = makeFingerprint(certificate);
        const isNew = !index.has(fingerprint);
        index.set(fingerprint, { certificate, filename, info });
        this.#filenameToHash.set(filename, fingerprint);
        debugLog(
          import_chalk6.default.magenta("CERT"),
          info.tbsCertificate.subjectFingerPrint,
          info.tbsCertificate.serialNumber,
          info.tbsCertificate.extensions?.authorityKeyIdentifier?.authorityCertIssuerFingerPrint
        );
        if (ready || isNew) {
          this.emit("certificateAdded", { store, certificate, fingerprint, filename });
        }
      } catch (err) {
        debugLog(`Walk files in folder ${folder} with file ${filename}`);
        debugLog(err);
      }
    });
    w.on("change", (changedPath) => {
      debugLog(import_chalk6.default.cyan(`change in folder ${folder}`), changedPath);
      try {
        const certificate = (0, import_node_opcua_crypto5.readCertificate)(changedPath);
        const newFingerprint = makeFingerprint(certificate);
        const oldHash = this.#filenameToHash.get(changedPath);
        if (oldHash && oldHash !== newFingerprint) {
          index.delete(oldHash);
        }
        index.set(newFingerprint, { certificate, filename: changedPath, info: exploreCertificateCached(certificate) });
        this.#filenameToHash.set(changedPath, newFingerprint);
        this.emit("certificateChange", { store, certificate, fingerprint: newFingerprint, filename: changedPath });
      } catch (err) {
        debugLog(`change event: failed to re-read ${changedPath}`, err);
      }
    });
    this.#watchers.push(w);
    this.#pendingUnrefs.add(unreffAll);
    w.on("ready", () => {
      ready = true;
      this.#pendingUnrefs.delete(unreffAll);
      unreffAll();
      debugLog("ready");
      debugLog([...index.keys()].map((k) => k.substring(0, 10)));
    });
  }
  // make sure that all crls have been processed.
  async #waitAndCheckCRLProcessingStatus() {
    return new Promise((resolve, _reject) => {
      if (this.#pendingCrlToProcess === 0) {
        setImmediate(resolve);
        return;
      }
      this.#onCrlProcessWaiters.push(resolve);
    });
  }
};

// packages/node-opcua-pki/lib/pki/toolbox_pfx.ts
var import_node_assert9 = __toESM(require("assert"));
var import_node_fs10 = __toESM(require("fs"));
var q2 = quote;
var n3 = makePath;
async function createPFX(options) {
  const { certificateFile, privateKeyFile, outputFile, passphrase = "", caCertificateFiles } = options;
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(certificateFile), `Certificate file does not exist: ${certificateFile}`);
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(privateKeyFile), `Private key file does not exist: ${privateKeyFile}`);
  let cmd = `pkcs12 -export`;
  cmd += ` -in ${q2(n3(certificateFile))}`;
  cmd += ` -inkey ${q2(n3(privateKeyFile))}`;
  if (caCertificateFiles) {
    for (const caFile of caCertificateFiles) {
      (0, import_node_assert9.default)(import_node_fs10.default.existsSync(caFile), `CA certificate file does not exist: ${caFile}`);
      cmd += ` -certfile ${q2(n3(caFile))}`;
    }
  }
  cmd += ` -out ${q2(n3(outputFile))}`;
  cmd += ` -passout pass:${passphrase}`;
  await execute_openssl(cmd, {});
}
async function extractCertificateFromPFX(options) {
  const { pfxFile, passphrase = "" } = options;
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q2(n3(pfxFile))} -clcerts -nokeys -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
async function extractPrivateKeyFromPFX(options) {
  const { pfxFile, passphrase = "" } = options;
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q2(n3(pfxFile))} -nocerts -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
async function extractCACertificatesFromPFX(options) {
  const { pfxFile, passphrase = "" } = options;
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q2(n3(pfxFile))} -cacerts -nokeys -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
async function extractAllFromPFX(options) {
  const [certificate, privateKey, caCertificates] = await Promise.all([
    extractCertificateFromPFX(options),
    extractPrivateKeyFromPFX(options),
    extractCACertificatesFromPFX(options)
  ]);
  return { certificate, privateKey, caCertificates };
}
async function convertPFXtoPEM(pfxFile, pemFile, passphrase = "") {
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q2(n3(pfxFile))} -out ${q2(n3(pemFile))} -nodes -passin pass:${passphrase}`;
  await execute_openssl(cmd, {});
}
async function dumpPFX(pfxFile, passphrase = "") {
  (0, import_node_assert9.default)(import_node_fs10.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q2(n3(pfxFile))} -info -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CertificateAuthority,
  CertificateManager,
  CertificateManagerState,
  Subject,
  VerificationStatus,
  adjustApplicationUri,
  adjustDate,
  convertPFXtoPEM,
  createPFX,
  dumpPFX,
  extractAllFromPFX,
  extractCACertificatesFromPFX,
  extractCertificateFromPFX,
  extractPrivateKeyFromPFX,
  findIssuerCertificateInChain,
  install_prerequisite,
  quote
});
